{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Antigenic analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Alphabet import generic_dna\n",
    "import pandas as pd\n",
    "import os\n",
    "import tempfile\n",
    "import sys\n",
    "import subprocess\n",
    "from scipy.stats import binom\n",
    "from ast import literal_eval\n",
    "from ggplot import *\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_to_summary(line_pattern,value):\n",
    "    lines = []\n",
    "    with open(\"./results.table.tsv\",'r') as results:\n",
    "        for line in results:\n",
    "            if line_pattern in line:\n",
    "                line = line_pattern +\"\\t\" + str(value)+\"\\n\"\n",
    "            lines.append(line)\n",
    "            \n",
    "    with open(\"./results.table.tsv\",'w') as output:\n",
    "        for l in lines:\n",
    "            output.write(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Align(headers_seqs, progpath, program='PROBCONS', musclegapopen=None):\n",
    "    \"\"\"Performs a multiple sequence alignment of two or more sequences.\n",
    "\n",
    "    By default, the protein sequences are aligned using PROBCONS.  This is\n",
    "        probably the most accurate alignment program.  However, it is\n",
    "        slow and consumes large amounts of memory if you are aligning\n",
    "        a very large number of sequences (typically if you are aligning\n",
    "        more than several hundred).  In that case, you may prefer to use\n",
    "        MUSCLE instead.  You can choose between the two with the 'program'\n",
    "        option.  If you decide to use MUSCLE, you can also align nucleotide\n",
    "        sequences with this program.\n",
    "\n",
    "    'headers_seqs' is a list specifying the names of the sequences that we\n",
    "        want to align.  Each entry is a 2-tuple '(head, seq)' where 'head' is\n",
    "        a header giving the sequence name and other information (might be empty)\n",
    "        and 'seq' is a string giving the protein sequence.  The list must have\n",
    "        at least 2 entries.\n",
    "\n",
    "    'progpath' should specify a directory containing the alignment program executable,\n",
    "        either PROBCONS or MUSCLE.  The PROBCONS executable is assumed to have\n",
    "        the name \"probcons\" in this directory.  The MUSCLE executable is assumed to\n",
    "        have the name \"muscle\" in this directory.\n",
    "\n",
    "    'program' specifies what program to use for the alignment.  By default, it is\n",
    "        \"PROBCONS\".  If you wish to use MUSCLE instead, set it to \"MUSCLE\".\n",
    "\n",
    "    'musclegapopen' sets the MUSCLE gap openining penalty to the specified\n",
    "        value. By default it is None, meaning we use the MUSCLE default penalty.\n",
    "        You can also set it to a number; for example -100 will lead to fewer gaps.\n",
    "\n",
    "    This executable is used to perform a multiple sequence alignment of the proteins\n",
    "        with the default settings of either PROBCONS or MUSCLE.  The returned variable is a\n",
    "        new list 'aligned_headers_seqs'.  Each entry is a 2-tuple '(head, aligned_seq)'.\n",
    "        'head' has the same meaning as on input (the sequence header) and\n",
    "        'aligned_seq' is the aligned sequence, with gaps inserted as '-'\n",
    "        as appropriate.  Therefore, all of the 'aligned_seq' entries in\n",
    "        'aligned_headers_seqs' are of the same length.  The entries in 'aligned_headers_seq'\n",
    "        are in the same order as in the input list 'headers_seqs'.\n",
    "    \"\"\"\n",
    "    if not (isinstance(headers_seqs, list) and len(headers_seqs) >= 2):\n",
    "        raise ValueError('header_seqs does not specify a list with at least two entries.')\n",
    "    if not os.path.isdir(progpath):\n",
    "        raise ValueError(\"Cannot find directory %s.\" % progpath)\n",
    "    if program == 'PROBCONS':\n",
    "        exe = os.path.abspath(\"%s/probcons\" % progpath) # the executable\n",
    "    elif program == 'MUSCLE':\n",
    "        exe = os.path.abspath(\"%s/muscle\" % progpath) # the executable\n",
    "    else:\n",
    "        raise ValueError(\"Invalid value of %s for 'program'.\" % (str(program)))\n",
    "    if not os.path.isfile(exe):\n",
    "        raise IOError(\"Cannot find executable at %s.\" % exe)\n",
    "    currdir = os.getcwd()\n",
    "    tempdir = tempfile.mkdtemp()\n",
    "    try:\n",
    "        # do stuff in a temporary directory\n",
    "        infile = \"%s/in.fasta\" % tempdir # input file\n",
    "        outfile = \"%s/out.fasta\" % tempdir # output file\n",
    "        WriteFASTA(headers_seqs, infile) # write sequences to the input file\n",
    "        if program == 'PROBCONS':\n",
    "            p = subprocess.Popen(\"%s %s\" % (exe, infile), shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE) # run ProbCons\n",
    "            (output, errors) = p.communicate()\n",
    "\n",
    "            open(outfile, 'w').write(output)\n",
    "        elif program == 'MUSCLE':\n",
    "            if musclegapopen != None:\n",
    "                p = subprocess.Popen(\"%s -gapopen %d -in %s -out %s\" % (exe, musclegapopen, infile, outfile), shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE) # run MUSCLE\n",
    "            else:\n",
    "                p = subprocess.Popen(\"%s -in %s -out %s\" % (exe, infile, outfile), shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE) # run MUSCLE\n",
    "            (output, errors) = p.communicate()\n",
    "        try:\n",
    "            aligned_headers_seqs = ReadFASTA(outfile)\n",
    "        except:\n",
    "            sys.stderr.write(\"Error getting alignment output, error of %s\" % errors)\n",
    "            raise\n",
    "    finally:\n",
    "        os.chdir(currdir) # return to the original directory\n",
    "        for file in os.listdir(tempdir):\n",
    "            os.remove(\"%s/%s\" % (tempdir, file)) # remove files from temporary directory\n",
    "        os.rmdir(tempdir) # remove temporary directory\n",
    "    if len(aligned_headers_seqs) != len(headers_seqs):\n",
    "        raise ValueError(\"Did not return the correct number of aligned sequences.\")\n",
    "    # put the aligned sequences in the same order as the input sequences\n",
    "    n = len(aligned_headers_seqs[0][1]) # length of aligned sequences\n",
    "    d = dict(aligned_headers_seqs)\n",
    "    aligned_headers_seqs = []\n",
    "    for (head, seq) in headers_seqs:\n",
    "        try:\n",
    "            alignedseq = d[head]\n",
    "        except KeyError:\n",
    "            raise ValueError(\"After alignment, the following header is missing: %s\" % head)\n",
    "        if len(alignedseq) != n:\n",
    "            open('errors.temp', 'w').write(errors)\n",
    "            raise ValueError(\"Aligned sequence %s is not of length %d: if you are using MUSCLE, you may be running out of memory.  Errors have been written to errors.temp.\" % (alignedseq, n))\n",
    "        if len(seq) > n:\n",
    "            open('errors.temp', 'w').write(errors)\n",
    "            raise ValueError(\"Unaligned seq %s is longer than aligned length of %d: if you are using MUSCLE, you many be running out of memory.  Errors have been written to errors.temp.\" % (seq, n))\n",
    "        aligned_headers_seqs.append((head, alignedseq))\n",
    "    return aligned_headers_seqs # return the aligned sequences\n",
    "\n",
    "\n",
    "def StripGapsToFirstSequence(aligned_headers_seqs):\n",
    "    \"\"\"Strips gaps from a reference sequence, and all corresponding alignments.\n",
    "\n",
    "    On input, 'aligned_headers_seqs' should be a set of two or more aligned sequences,\n",
    "        as would be returned by 'Align'.\n",
    "\n",
    "    The first sequence in this alignment is taken to correspond to the reference sequence.\n",
    "        The returned variable is a list similar to aligned_headers_seqs, but with\n",
    "        all positions corresponding to gaps in this reference sequence stripped away.\n",
    "        All gaps ('-') characters are removed from this reference sequence.  In addition,\n",
    "        in all other aligned sequences in 'aligned_headers_seqs', every character at\n",
    "        the same position as a gap in the reference sequence is removed.  Therefore,\n",
    "        at the end of this procedure, all of the alignments have the same length\n",
    "        as the reference sequence with its gaps stripped away.  The headers are\n",
    "        unchanged.  The order of sequences in this stripped alignment is also\n",
    "        unchanged.\n",
    "\n",
    "    >>> StripGapsToFirstSequence([('s1', '-AT-A-GC'), ('s2', 'AAT-TAGC'), ('s3', '--T-A-GC')])\n",
    "    [('s1', 'ATAGC'), ('s2', 'ATTGC'), ('s3', '-TAGC')]\n",
    "    \"\"\"\n",
    "    if not (isinstance(aligned_headers_seqs, list) and len(aligned_headers_seqs) >= 2):\n",
    "        raise ValueError(\"aligned_headers_seqs does not specify at least two aligned sequences.\")\n",
    "    (ref_head, ref_seq) = aligned_headers_seqs[0]\n",
    "    non_strip_positions = [] # positions not to strip away\n",
    "    stripped_ref_seq = []\n",
    "    for i in range(len(ref_seq)):\n",
    "        r = ref_seq[i]\n",
    "        if r != '-':\n",
    "            non_strip_positions.append(i)\n",
    "            stripped_ref_seq.append(r)\n",
    "    stripped_headers_seqs = [(ref_head, ''.join(stripped_ref_seq))]\n",
    "    for (ihead, iseq) in aligned_headers_seqs[1 : ]:\n",
    "        istrippedseq = ''.join([iseq[i] for i in non_strip_positions])\n",
    "        stripped_headers_seqs.append((ihead, istrippedseq))\n",
    "    return stripped_headers_seqs\n",
    "\n",
    "\n",
    "def WriteFASTA(headers_seqs, filename, writable_file=False):\n",
    "    \"\"\"Writes sequences to a FASTA file.\n",
    "\n",
    "    'headers_seqs' is a list of 2-tuples specifying sequences and their\n",
    "        corresponding headers.  Each entry is the 2-tuple '(header, seq)'\n",
    "        where 'header' is a string giving the header (without the leading \">\"),\n",
    "        and 'seq' is the corresponding sequence.\n",
    "\n",
    "    'filename' is a string that specifies the name of the file to which the\n",
    "        headers and sequences should be written.  If this file already exists,\n",
    "        it is overwritten.\n",
    "\n",
    "    'writable_file' is a Boolean switch specifying that rather than 'filename'\n",
    "        giving a string specifying the name of a file to which the sequences\n",
    "        should be written, it instead specifies a writable file object to which\n",
    "        the sequences should be written.\n",
    "\n",
    "    The sequences are written to the file in the same order that they are specified\n",
    "        in 'headers_seqs'.\n",
    "    \"\"\"\n",
    "    assert isinstance(writable_file, bool)\n",
    "    if writable_file:\n",
    "        f = filename\n",
    "    else:\n",
    "        f = open(filename, 'w')\n",
    "    for (header, seq) in headers_seqs:\n",
    "        f.write(\">%s\\n%s\\n\" % (header, seq))\n",
    "    if not writable_file:\n",
    "        f.close()\n",
    "\n",
    "\n",
    "\n",
    "def ReadFASTA(fastafile):\n",
    "    \"\"\"Reads sequences from a FASTA file.\n",
    "\n",
    "    'fastafile' should specify the name of a FASTA file.\n",
    "\n",
    "    This function reads all sequences from the FASTA file.  It returns the\n",
    "        list 'headers_seqs'.  This list is composed of a 2-tuple '(header, seq)'\n",
    "        for every sequence entry in FASTA file.  'header' is the header for\n",
    "        a sequence, with the leading \">\" and any trailing spaces removed. 'seq'\n",
    "        is the corresponding sequence.\n",
    "    \"\"\"\n",
    "    lines = open(fastafile).readlines()\n",
    "    headers_seqs = []\n",
    "    header = None\n",
    "    seq = []\n",
    "    for line in lines:\n",
    "        if line[0] == '>':\n",
    "            if (not header) and (not seq):\n",
    "                pass # first sequence in file\n",
    "            elif header and not seq:\n",
    "                raise ValueError(\"Empty sequence for %s\" % header)\n",
    "            elif seq and not header:\n",
    "                raise ValueError(\"File does not begin with header.\")\n",
    "            else:\n",
    "                seq = ''.join(seq)\n",
    "                seq = seq.replace(' ', '')\n",
    "                headers_seqs.append((header, seq))\n",
    "            header = line.strip()[1 : ]\n",
    "            seq = []\n",
    "        else:\n",
    "            seq.append(line.strip())\n",
    "    if (not header) and (not seq):\n",
    "        pass # first sequence in file\n",
    "    elif header and not seq:\n",
    "        raise ValueError(\"Empty sequence for %s\" % header)\n",
    "    elif seq and not header:\n",
    "        raise ValueError(\"File does not begin with header.\")\n",
    "    else:\n",
    "        seq = ''.join(seq)\n",
    "        seq = seq.replace(' ', '')\n",
    "        headers_seqs.append((header, seq))\n",
    "    return headers_seqs\n",
    "\n",
    "\n",
    "def GetCorrespondingResidue(seqs, i):\n",
    "    \"\"\"Gets the corresponding residue number for two aligned sequences.\n",
    "\n",
    "    *seqs* is a set of two aligned sequences as *(head, sequence)* 2-tuples.\n",
    "\n",
    "    *i* is the number of a residue in sequential numbering of *seqs[0]*\n",
    "    without considering any of the gaps induced by alignment, in 1, 2, ...\n",
    "    numbering.\n",
    "\n",
    "    Returns the number of the residue in sequential numbering of *seqs[1]*\n",
    "    without considering any of the gaps induced by alignment in 1, 2, ...\n",
    "    numbering. Returns *None* if residue *i* of *seqs[0]* aligns with a\n",
    "    gap in *seqs[1]*.\n",
    "    \"\"\"\n",
    "    assert len(seqs) == 2\n",
    "    s1 = seqs[0][1]\n",
    "    s2 = seqs[1][1]\n",
    "    assert len(s1) == len(s2)\n",
    "    assert 1 <= i <= len(s1)\n",
    "    s1index = s2index = 0\n",
    "    for j in range(len(s1)):\n",
    "        if s1[j] != '-':\n",
    "            s1index += 1\n",
    "        if s2[j] != '-':\n",
    "            s2index += 1\n",
    "        if s1index == i:\n",
    "            if s2[j] == '-':\n",
    "                return None\n",
    "            else:\n",
    "                return s2index\n",
    "\n",
    "\n",
    "def find_aligments(aligner,aligner_path,seq,pos): # The positions will be given as base 0 and adjusted to match the convention (base 1) in the funciton\n",
    "    \"\"\"Main body of script.\"\"\"\n",
    "    print(\"\\nBeginning execution of HA_numbering script.\")\n",
    "\n",
    "    # parse arguments\n",
    "    if aligner == 'probcons':\n",
    "        alignerpath = aligner_path # redundant\n",
    "        if not os.path.isdir(alignerpath):\n",
    "            raise IOError(\"The directory of %s specified by probconspath does not exist.\" % (alignerpath))\n",
    "        prog = 'PROBCONS'\n",
    "    elif aligner == 'muscle':\n",
    "        alignerpath = aligner_path # redundant\n",
    "        if not os.path.isdir(alignerpath):\n",
    "            raise IOError(\"The directory of %s specified by musclepath does not exist.\" % (alignerpath))\n",
    "        prog = 'MUSCLE'\n",
    "    else:\n",
    "        raise IOError(\"First argument does not specify probcons or muscle\")\n",
    "    ha_sequence = str(seq) # just get the sequence\n",
    "    sites = pos\n",
    "    # Define sequences and their numbering conversions.\n",
    "    # The sequences are in seq_d and keyed by PDB code.\n",
    "    # The number conversions from sequential numbering of these sequences to\n",
    "    # th number labels are in label_d and keyed by PDB codes.\n",
    "    seq_names = ['4HMG', '4JTV']\n",
    "    seq_d = {}\n",
    "    label_d = {}\n",
    "    seq_4hmg_a = \\\n",
    "        'QDLPGNDNSTATLCLGHHAVPNGTLVKTITDDQIEVTNATELVQSSSTGKICNNPHRILDGIDCTLIDALLGDPHCDVFQ' +\\\n",
    "        'NETWDLFVERSKAFSNCYPYDVPDYASLRSLVASSGTLEFITEGFTWTGVTQNGGSNACKRGPGSGFFSRLNWLTKSGST' +\\\n",
    "        'YPVLNVTMPNNDNFDKLYIWGIHHPSTNQEQTSLYVQASGRVTVSTRRSQQTIIPNIGSRPWVRGQSSRISIYWTIVKPG' +\\\n",
    "        'DVLVINSNGNLIAPRGYFKMRTGKSSIMRSDAPIDTCISECITPNGSIPNDKPFQNVNKITYGACPKYVKQNTLKLATGM' +\\\n",
    "        'RNVPEKQT'\n",
    "    seq_4hmg_b = \\\n",
    "        'GLFGAIAGFIENGWEGMIDGWYGFRHQNSEGTGQAADLKSTQAAIDQINGKLNRVIEKTNEKFHQIEKEFSEVEGRIQDL' +\\\n",
    "        'EKYVEDTKIDLWSYNAELLVALENQHTIDLTDSEMNKLFEKTRRQLRENAEEMGNGCFKIYHKCDNACIESIRNGTYDHD' +\\\n",
    "        'VYRDEALNNRFQIKG'\n",
    "    seq_d['4HMG'] = seq_4hmg_a + seq_4hmg_b\n",
    "    label_d['4HMG'] = dict([(i + 1, '%d in HA1' % (i + 1)) for i in range(len(seq_4hmg_a))] + [(len(seq_4hmg_a) + i + 1, '%d in HA2' % (i + 1)) for i in range(len(seq_4hmg_b))])\n",
    "    assert len(seq_d['4HMG']) == len(label_d['4HMG'])\n",
    "    seq_4jtv_a = \\\n",
    "        'DTLCIGYHANNSTDTVDTVLEKNVTVTHSVNLLEDKHNGKLCKLRGVAPLHLGKCNIAGWILGNPECESLSTASSWSYIV' +\\\n",
    "        'ETPSSDNGTCYPGDFIDYEELREQLSSVSSFERFEIFPKTSSWPNHDSNKGVTAACPHAGAKSFYKNLIWLVKKGNSYPK' +\\\n",
    "        'LSKSYINDKGKEVLVLWGIHHPSTSADQQSLYQNADTYVFVGSSRYSKKFKPEIAIRPKVRDQEGRMNYYWTLVEPGDKI' +\\\n",
    "        'TFEATGNLVVPRYAFAMERNAGSGIIISDTPVHDCNTTCQTPKGAINTSLPFQNIHPITIGKCPKYVKSTKLRLATGLRN' +\\\n",
    "        'I'\n",
    "    seq_4jtv_b = \\\n",
    "        'GLFGAIAGFIEGGWTGMVDGWYGYHHQNEQGSGYAADLKSTQNAIDEITNKVNSVIEKMNTQFTAVGKEFNHLEKRIENL' +\\\n",
    "        'NKKVDDGFLDIWTYNAELLVLLENERTLDYHDSNVKNLYEKVRSQLKNNAKEIGNGCFEFYHKCDNTCMESVKNGTYDYP' +\\\n",
    "        'KY'\n",
    "    seq_d['4JTV'] = seq_4jtv_a + seq_4jtv_b\n",
    "    label_d['4JTV'] = dict([(i + 1, '%d of HA1' % (i + 7)) for i in range(len(seq_4jtv_a))] + [(len(seq_4jtv_a) + i + 1, '%d of HA2' % (i + 1)) for i in range(len(seq_4jtv_b))])\n",
    "    assert len(seq_d['4JTV']) == len(label_d['4JTV'])\n",
    "\n",
    "    # make alignments\n",
    "    print(\"Making %s alignments...\" % prog)\n",
    "    alignments = {}\n",
    "    for seqname in seq_names:\n",
    "        alignments[seqname] = Align([('seq', ha_sequence), (seqname, seq_d[seqname])], alignerpath, prog)\n",
    "    print(\"Alignments complete.\\n\\nHere are the corresponding residue numbers:\")\n",
    "    #print(sites)\n",
    "    out_hmg=[]\n",
    "    out_jtv = []\n",
    "    for site in sites:\n",
    "        #site=site+1 # to correct for difference in base numbering\n",
    "        if not (1 <= site <= len(ha_sequence)):\n",
    "            raise ValueError(\"site %d is outside the valid range for sequential numbering of ha_sequence starting at 1.\" % site)\n",
    "        sitestring = ['\\nResidue %s%d in sequential numbering of your HA sequence corresponds to:' % (ha_sequence[site - 1], site)]\n",
    "        for seqname in seq_names:\n",
    "            i = GetCorrespondingResidue(alignments[seqname], site)\n",
    "            if i == None:\n",
    "                sitestring.append('  * an alignment gap in %s' % seqname)\n",
    "            else:\n",
    "                if not (1 <= i <= len(seq_d[seqname])):\n",
    "                    raise ValueError(\"Invalid corrresponding residue for %s -- something is wrong with this program\" % seqname)\n",
    "                aa = seq_d[seqname][i - 1]\n",
    "\n",
    "                sitestring.append('  * %s%s in %s' % (aa, label_d[seqname][i], seqname))\n",
    "        print( '\\n'.join(sitestring))\n",
    "        out_hmg.append(sitestring[1][4:-7])\n",
    "        out_jtv.append(sitestring[2][4:-7])\n",
    "    \n",
    "    return([out_hmg,out_jtv])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting the  antigenic regions\n",
    "\n",
    "## H1N1\n",
    "Here I have copied the antigenic sites listed in Xu et al. 2010. I am using a derivation of the Bloom lab script to align them to the PBD protien. The script requires base 1 numbering\n",
    "## H3N2\n",
    "I am also reading in H3 antigenic regions noted in Lee et al. 2004 with a special column noting if the site was also mentioned in Wiley et al. 1981 and/or Smith et al. 2004.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Antigenic site strings  \"site : [ [list of sequences], [List of lists of antigenic positions in that sequence]]\n",
    "H1N1 = {'Sa' : [[\"PNHDS\",\"KGNSYPKLSKS\"],[[1,2],[1,2,3,4,6,7,8,9,10,11]]] ,\n",
    "        'Sb':[[\"TSADQQSLYQNA\"],[[1,2,3,4,5,6,7,8,9,10,11,12]]],\n",
    "        'Ca' : [[\"PHAGAKSF\",\"INDKGK\",\"SSRYSK\",\"KVRDQ\",\"EPGDKI\"],[[1,2,3,4,5,6],[1,2,3,4,5],[1,2,3],[3,4],[1,2,3]]],\n",
    "        'Cb' : [[\"LSTASSWSY\"], [[1,2,3,4,5,6]]]}\n",
    "\n",
    "H3 = pd.read_csv(\"../data/reference/H3N2_antigenic_sites.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This chunck does the H1N1 aligments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PHAGAKSF\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P215 in HA1 in 4HMG\n",
      "  * P143 of HA1 in 4JTV\n",
      "\n",
      "Residue H2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N216 in HA1 in 4HMG\n",
      "  * H144 of HA1 in 4JTV\n",
      "\n",
      "Residue A3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I217 in HA1 in 4HMG\n",
      "  * A145 of HA1 in 4JTV\n",
      "\n",
      "Residue G4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G218 in HA1 in 4HMG\n",
      "  * G146 of HA1 in 4JTV\n",
      "\n",
      "Residue A5 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S219 in HA1 in 4HMG\n",
      "  * A147 of HA1 in 4JTV\n",
      "\n",
      "Residue K6 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R220 in HA1 in 4HMG\n",
      "  * K148 of HA1 in 4JTV\n",
      "INDKGK\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I245 in HA1 in 4HMG\n",
      "  * I172 of HA1 in 4JTV\n",
      "\n",
      "Residue N2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N246 in HA1 in 4HMG\n",
      "  * N173 of HA1 in 4JTV\n",
      "\n",
      "Residue D3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S247 in HA1 in 4HMG\n",
      "  * D174 of HA1 in 4JTV\n",
      "\n",
      "Residue K4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N248 in HA1 in 4HMG\n",
      "  * K175 of HA1 in 4JTV\n",
      "\n",
      "Residue G5 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G249 in HA1 in 4HMG\n",
      "  * G176 of HA1 in 4JTV\n",
      "SSRYSK\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * C305 in HA1 in 4HMG\n",
      "  * S209 of HA1 in 4JTV\n",
      "\n",
      "Residue S2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P306 in HA1 in 4HMG\n",
      "  * S210 of HA1 in 4JTV\n",
      "\n",
      "Residue R3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K307 in HA1 in 4HMG\n",
      "  * R211 of HA1 in 4JTV\n",
      "KVRDQ\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R127 in HA2 in 4HMG\n",
      "  * R227 of HA1 in 4JTV\n",
      "\n",
      "Residue D4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E128 in HA2 in 4HMG\n",
      "  * D228 of HA1 in 4JTV\n",
      "EPGDKI\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K238 in HA1 in 4HMG\n",
      "  * E241 of HA1 in 4JTV\n",
      "\n",
      "Residue P2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P239 in HA1 in 4HMG\n",
      "  * P242 of HA1 in 4JTV\n",
      "\n",
      "Residue G3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G240 in HA1 in 4HMG\n",
      "  * G243 of HA1 in 4JTV\n",
      "LSTASSWSY\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L154 in HA1 in 4HMG\n",
      "  * L76 of HA1 in 4JTV\n",
      "\n",
      "Residue S2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T155 in HA1 in 4HMG\n",
      "  * S77 of HA1 in 4JTV\n",
      "\n",
      "Residue T3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K156 in HA1 in 4HMG\n",
      "  * T78 of HA1 in 4JTV\n",
      "\n",
      "Residue A4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S157 in HA1 in 4HMG\n",
      "  * A79 of HA1 in 4JTV\n",
      "\n",
      "Residue S5 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G158 in HA1 in 4HMG\n",
      "  * S80 of HA1 in 4JTV\n",
      "\n",
      "Residue S6 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S159 in HA1 in 4HMG\n",
      "  * S81 of HA1 in 4JTV\n",
      "PNHDS\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P169 in HA1 in 4HMG\n",
      "  * P130 of HA1 in 4JTV\n",
      "\n",
      "Residue N2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N170 in HA1 in 4HMG\n",
      "  * N131 of HA1 in 4JTV\n",
      "KGNSYPKLSKS\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S157 in HA1 in 4HMG\n",
      "  * K160 of HA1 in 4JTV\n",
      "\n",
      "Residue G2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G158 in HA1 in 4HMG\n",
      "  * G161 of HA1 in 4JTV\n",
      "\n",
      "Residue N3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S159 in HA1 in 4HMG\n",
      "  * N162 of HA1 in 4JTV\n",
      "\n",
      "Residue S4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T160 in HA1 in 4HMG\n",
      "  * S163 of HA1 in 4JTV\n",
      "\n",
      "Residue P6 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P162 in HA1 in 4HMG\n",
      "  * P165 of HA1 in 4JTV\n",
      "\n",
      "Residue K7 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V163 in HA1 in 4HMG\n",
      "  * K166 of HA1 in 4JTV\n",
      "\n",
      "Residue L8 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L164 in HA1 in 4HMG\n",
      "  * L167 of HA1 in 4JTV\n",
      "\n",
      "Residue S9 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N165 in HA1 in 4HMG\n",
      "  * S168 of HA1 in 4JTV\n",
      "\n",
      "Residue K10 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V166 in HA1 in 4HMG\n",
      "  * K169 of HA1 in 4JTV\n",
      "\n",
      "Residue S11 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T167 in HA1 in 4HMG\n",
      "  * S170 of HA1 in 4JTV\n",
      "TSADQQSLYQNA\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T1 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T187 in HA1 in 4HMG\n",
      "  * T190 of HA1 in 4JTV\n",
      "\n",
      "Residue S2 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N188 in HA1 in 4HMG\n",
      "  * S191 of HA1 in 4JTV\n",
      "\n",
      "Residue A3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q189 in HA1 in 4HMG\n",
      "  * A192 of HA1 in 4JTV\n",
      "\n",
      "Residue D4 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E190 in HA1 in 4HMG\n",
      "  * D193 of HA1 in 4JTV\n",
      "\n",
      "Residue Q5 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q191 in HA1 in 4HMG\n",
      "  * Q194 of HA1 in 4JTV\n",
      "\n",
      "Residue Q6 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T192 in HA1 in 4HMG\n",
      "  * Q195 of HA1 in 4JTV\n",
      "\n",
      "Residue S7 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S193 in HA1 in 4HMG\n",
      "  * S196 of HA1 in 4JTV\n",
      "\n",
      "Residue L8 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L194 in HA1 in 4HMG\n",
      "  * L197 of HA1 in 4JTV\n",
      "\n",
      "Residue Y9 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Y195 in HA1 in 4HMG\n",
      "  * Y198 of HA1 in 4JTV\n",
      "\n",
      "Residue Q10 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V196 in HA1 in 4HMG\n",
      "  * Q199 of HA1 in 4JTV\n",
      "\n",
      "Residue N11 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q197 in HA1 in 4HMG\n",
      "  * N200 of HA1 in 4JTV\n",
      "\n",
      "Residue A12 in sequential numbering of your HA sequence corresponds to:\n",
      "  * A198 in HA1 in 4HMG\n",
      "  * A201 of HA1 in 4JTV\n",
      "['P143 of HA1 ', 'H144 of HA1 ', 'A145 of HA1 ', 'G146 of HA1 ', 'A147 of HA1 ', 'K148 of HA1 ', 'I172 of HA1 ', 'N173 of HA1 ', 'D174 of HA1 ', 'K175 of HA1 ', 'G176 of HA1 ', 'S209 of HA1 ', 'S210 of HA1 ', 'R211 of HA1 ', 'R227 of HA1 ', 'D228 of HA1 ', 'E241 of HA1 ', 'P242 of HA1 ', 'G243 of HA1 ', 'L76 of HA1 ', 'S77 of HA1 ', 'T78 of HA1 ', 'A79 of HA1 ', 'S80 of HA1 ', 'S81 of HA1 ', 'P130 of HA1 ', 'N131 of HA1 ', 'K160 of HA1 ', 'G161 of HA1 ', 'N162 of HA1 ', 'S163 of HA1 ', 'P165 of HA1 ', 'K166 of HA1 ', 'L167 of HA1 ', 'S168 of HA1 ', 'K169 of HA1 ', 'S170 of HA1 ', 'T190 of HA1 ', 'S191 of HA1 ', 'A192 of HA1 ', 'D193 of HA1 ', 'Q194 of HA1 ', 'Q195 of HA1 ', 'S196 of HA1 ', 'L197 of HA1 ', 'Y198 of HA1 ', 'Q199 of HA1 ', 'N200 of HA1 ', 'A201 of HA1 ']\n"
     ]
    }
   ],
   "source": [
    "antigenic = []\n",
    "for site in H1N1:\n",
    "    i = 0\n",
    "    for seq in H1N1[site][0]:\n",
    "        print(seq)\n",
    "        nice = find_aligments('muscle',\"/Users/jt/muscle3.8.31/\",seq,H1N1[site][1][i])\n",
    "        for pos in nice[1]:\n",
    "            antigenic.append(pos)\n",
    "        i+=1\n",
    "\n",
    "print(antigenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Anitgenic analysis\n",
    "\n",
    "Here we will be reading in the quality snv from our analysis. We will focus on Nonsynonymous variants in HA for obvious reasons And I am selecting only minor variants.\n",
    "\n",
    "Note : Here the reference amino acid is the sample consensus not the plasmid control (as is the case for the nucleotide)\n",
    "\n",
    "For each ciruculating strain we will for each variant we will find the 4HMG and 4JTV sequence location. These will be used to identify variants in antigenic regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (36) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "snv = pd.read_csv(\"/Users/jt/Documents/Analysis/HIVE/data/processed/secondary/qual.snv.csv\",converters={\"AA_pos\": literal_eval})\n",
    "snv = snv.loc[snv['freq.var']<0.5]\n",
    "snv = snv.loc[(snv.SPECID != \"HS1530\") & (snv.SPECID !=\"MH8137\") &(snv.SPECID !=\"MH8390\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>run</th>\n",
       "      <th>X1.x</th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Id</th>\n",
       "      <th>LAURING_ID</th>\n",
       "      <th>dup</th>\n",
       "      <th>MapQ</th>\n",
       "      <th>Phred</th>\n",
       "      <th>Read_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>log_copy_num</th>\n",
       "      <th>gc_ul</th>\n",
       "      <th>HIGHSD</th>\n",
       "      <th>sequenced</th>\n",
       "      <th>home_collected</th>\n",
       "      <th>snv_qualified</th>\n",
       "      <th>X1</th>\n",
       "      <th>sample_coverage</th>\n",
       "      <th>coverage</th>\n",
       "      <th>class_factor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>perth</td>\n",
       "      <td>1</td>\n",
       "      <td>125</td>\n",
       "      <td>1102</td>\n",
       "      <td>1102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.672713</td>\n",
       "      <td>37.195615</td>\n",
       "      <td>63.517421</td>\n",
       "      <td>...</td>\n",
       "      <td>6.486882</td>\n",
       "      <td>219156.158420</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>403</td>\n",
       "      <td>29478.353879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>perth</td>\n",
       "      <td>3</td>\n",
       "      <td>216</td>\n",
       "      <td>1102</td>\n",
       "      <td>1102</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.477381</td>\n",
       "      <td>36.948810</td>\n",
       "      <td>78.584524</td>\n",
       "      <td>...</td>\n",
       "      <td>6.486882</td>\n",
       "      <td>219156.158420</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>403</td>\n",
       "      <td>29478.353879</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>275</td>\n",
       "      <td>cali09</td>\n",
       "      <td>1</td>\n",
       "      <td>108</td>\n",
       "      <td>1112</td>\n",
       "      <td>1112</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.383436</td>\n",
       "      <td>37.535106</td>\n",
       "      <td>61.826005</td>\n",
       "      <td>...</td>\n",
       "      <td>6.044844</td>\n",
       "      <td>79198.379568</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>17796.126556</td>\n",
       "      <td>27052.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>328</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>1</td>\n",
       "      <td>173</td>\n",
       "      <td>1112</td>\n",
       "      <td>1112</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.851569</td>\n",
       "      <td>37.230859</td>\n",
       "      <td>62.804341</td>\n",
       "      <td>...</td>\n",
       "      <td>6.044844</td>\n",
       "      <td>79198.379568</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>54</td>\n",
       "      <td>16560.351942</td>\n",
       "      <td>14463.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>423</td>\n",
       "      <td>perth</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>1121</td>\n",
       "      <td>1121</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38.713415</td>\n",
       "      <td>37.501220</td>\n",
       "      <td>62.269512</td>\n",
       "      <td>...</td>\n",
       "      <td>5.060653</td>\n",
       "      <td>8213.445551</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>409</td>\n",
       "      <td>23610.190132</td>\n",
       "      <td>40503.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>507</th>\n",
       "      <td>508</td>\n",
       "      <td>perth</td>\n",
       "      <td>13452</td>\n",
       "      <td>111</td>\n",
       "      <td>1121</td>\n",
       "      <td>1121</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.534214</td>\n",
       "      <td>37.391014</td>\n",
       "      <td>65.533185</td>\n",
       "      <td>...</td>\n",
       "      <td>5.060653</td>\n",
       "      <td>8213.445551</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>409</td>\n",
       "      <td>23610.190132</td>\n",
       "      <td>22358.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699</th>\n",
       "      <td>700</td>\n",
       "      <td>perth_2</td>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>1122</td>\n",
       "      <td>1122</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.909170</td>\n",
       "      <td>37.092561</td>\n",
       "      <td>63.629758</td>\n",
       "      <td>...</td>\n",
       "      <td>4.315119</td>\n",
       "      <td>1475.675381</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>465</td>\n",
       "      <td>23418.068719</td>\n",
       "      <td>25880.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>876</th>\n",
       "      <td>877</td>\n",
       "      <td>perth_2</td>\n",
       "      <td>2</td>\n",
       "      <td>79</td>\n",
       "      <td>1125</td>\n",
       "      <td>1125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.608527</td>\n",
       "      <td>37.136628</td>\n",
       "      <td>61.093346</td>\n",
       "      <td>...</td>\n",
       "      <td>4.851537</td>\n",
       "      <td>5074.685473</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>466</td>\n",
       "      <td>19125.357084</td>\n",
       "      <td>14718.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904</th>\n",
       "      <td>905</td>\n",
       "      <td>perth_2</td>\n",
       "      <td>3</td>\n",
       "      <td>103</td>\n",
       "      <td>1125</td>\n",
       "      <td>1125</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.486762</td>\n",
       "      <td>37.289206</td>\n",
       "      <td>63.089613</td>\n",
       "      <td>...</td>\n",
       "      <td>4.851537</td>\n",
       "      <td>5074.685473</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>466</td>\n",
       "      <td>19125.357084</td>\n",
       "      <td>17276.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1148</th>\n",
       "      <td>1149</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13453</td>\n",
       "      <td>23</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.472222</td>\n",
       "      <td>37.330409</td>\n",
       "      <td>61.217836</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>26321.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1152</th>\n",
       "      <td>1153</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13461</td>\n",
       "      <td>42</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40.915110</td>\n",
       "      <td>37.259762</td>\n",
       "      <td>63.876061</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>25227.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1160</th>\n",
       "      <td>1161</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13455</td>\n",
       "      <td>25</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.772093</td>\n",
       "      <td>37.200000</td>\n",
       "      <td>64.624806</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>26539.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1164</th>\n",
       "      <td>1165</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13448</td>\n",
       "      <td>5</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.438679</td>\n",
       "      <td>37.498428</td>\n",
       "      <td>60.407233</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>23388.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1193</th>\n",
       "      <td>1194</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13488</td>\n",
       "      <td>113</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.549844</td>\n",
       "      <td>37.414330</td>\n",
       "      <td>65.526480</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>23484.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1197</th>\n",
       "      <td>1198</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13490</td>\n",
       "      <td>116</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.793333</td>\n",
       "      <td>37.235000</td>\n",
       "      <td>61.713333</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>22687.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1208</th>\n",
       "      <td>1209</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13483</td>\n",
       "      <td>106</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.820922</td>\n",
       "      <td>37.320922</td>\n",
       "      <td>62.187943</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>22622.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1212</th>\n",
       "      <td>1213</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13489</td>\n",
       "      <td>115</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.679758</td>\n",
       "      <td>37.332326</td>\n",
       "      <td>61.658610</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>23975.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1232</th>\n",
       "      <td>1233</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13508</td>\n",
       "      <td>161</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.734899</td>\n",
       "      <td>37.399329</td>\n",
       "      <td>63.546980</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>14236.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1236</th>\n",
       "      <td>1237</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13511</td>\n",
       "      <td>169</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.740484</td>\n",
       "      <td>37.155709</td>\n",
       "      <td>61.806228</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>13880.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1244</th>\n",
       "      <td>1245</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13525</td>\n",
       "      <td>214</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.594235</td>\n",
       "      <td>37.485588</td>\n",
       "      <td>61.647450</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>16145.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1247</th>\n",
       "      <td>1248</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13532</td>\n",
       "      <td>228</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.147982</td>\n",
       "      <td>37.172646</td>\n",
       "      <td>63.387892</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>15536.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1254</th>\n",
       "      <td>1255</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13531</td>\n",
       "      <td>227</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.549528</td>\n",
       "      <td>37.242925</td>\n",
       "      <td>64.320755</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>17207.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1260</th>\n",
       "      <td>1261</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13526</td>\n",
       "      <td>215</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.537209</td>\n",
       "      <td>37.320930</td>\n",
       "      <td>61.581395</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>16728.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1263</th>\n",
       "      <td>1264</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13528</td>\n",
       "      <td>219</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.860520</td>\n",
       "      <td>37.442080</td>\n",
       "      <td>61.669031</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>16941.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266</th>\n",
       "      <td>1267</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13535</td>\n",
       "      <td>235</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>40.905983</td>\n",
       "      <td>37.457265</td>\n",
       "      <td>64.395299</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>17076.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1279</th>\n",
       "      <td>1280</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13523</td>\n",
       "      <td>211</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.588670</td>\n",
       "      <td>36.928571</td>\n",
       "      <td>63.371921</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>15210.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1283</th>\n",
       "      <td>1284</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13554</td>\n",
       "      <td>282</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.795359</td>\n",
       "      <td>37.251055</td>\n",
       "      <td>58.856540</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>15795.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1293</th>\n",
       "      <td>1294</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13546</td>\n",
       "      <td>267</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.602537</td>\n",
       "      <td>37.456660</td>\n",
       "      <td>59.617336</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>14461.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297</th>\n",
       "      <td>1298</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>13557</td>\n",
       "      <td>292</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41.448203</td>\n",
       "      <td>37.427061</td>\n",
       "      <td>65.701903</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>55</td>\n",
       "      <td>15518.761124</td>\n",
       "      <td>17470.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1300</th>\n",
       "      <td>1301</td>\n",
       "      <td>cali09</td>\n",
       "      <td>2</td>\n",
       "      <td>248</td>\n",
       "      <td>1137</td>\n",
       "      <td>1137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>38.980501</td>\n",
       "      <td>37.504178</td>\n",
       "      <td>60.540390</td>\n",
       "      <td>...</td>\n",
       "      <td>4.711363</td>\n",
       "      <td>3674.806892</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>5</td>\n",
       "      <td>12298.362302</td>\n",
       "      <td>11155.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26147</th>\n",
       "      <td>26148</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>6</td>\n",
       "      <td>120</td>\n",
       "      <td>MH8592_A</td>\n",
       "      <td>MH8592</td>\n",
       "      <td>A</td>\n",
       "      <td>30.064327</td>\n",
       "      <td>37.286550</td>\n",
       "      <td>58.122807</td>\n",
       "      <td>...</td>\n",
       "      <td>6.547221</td>\n",
       "      <td>251821.707282</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>125</td>\n",
       "      <td>23957.041996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26412</th>\n",
       "      <td>26413</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>13581</td>\n",
       "      <td>70</td>\n",
       "      <td>MH8690_B</td>\n",
       "      <td>MH8690</td>\n",
       "      <td>B</td>\n",
       "      <td>41.493766</td>\n",
       "      <td>37.441397</td>\n",
       "      <td>67.558603</td>\n",
       "      <td>...</td>\n",
       "      <td>4.267601</td>\n",
       "      <td>1322.734812</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>202</td>\n",
       "      <td>3041.026127</td>\n",
       "      <td>3309.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26443</th>\n",
       "      <td>26444</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>5</td>\n",
       "      <td>104</td>\n",
       "      <td>MH8690_B</td>\n",
       "      <td>MH8690</td>\n",
       "      <td>B</td>\n",
       "      <td>41.378205</td>\n",
       "      <td>37.435897</td>\n",
       "      <td>62.705128</td>\n",
       "      <td>...</td>\n",
       "      <td>4.267601</td>\n",
       "      <td>1322.734812</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>202</td>\n",
       "      <td>3041.026127</td>\n",
       "      <td>3632.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26478</th>\n",
       "      <td>26479</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>MH8692_A</td>\n",
       "      <td>MH8692</td>\n",
       "      <td>A</td>\n",
       "      <td>41.720973</td>\n",
       "      <td>37.457751</td>\n",
       "      <td>61.563526</td>\n",
       "      <td>...</td>\n",
       "      <td>4.418474</td>\n",
       "      <td>1872.171013</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>132</td>\n",
       "      <td>21576.810909</td>\n",
       "      <td>43706.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26484</th>\n",
       "      <td>26485</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>MH8692_A</td>\n",
       "      <td>MH8692</td>\n",
       "      <td>A</td>\n",
       "      <td>41.283700</td>\n",
       "      <td>37.460801</td>\n",
       "      <td>63.990976</td>\n",
       "      <td>...</td>\n",
       "      <td>4.418474</td>\n",
       "      <td>1872.171013</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>132</td>\n",
       "      <td>21576.810909</td>\n",
       "      <td>35421.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26488</th>\n",
       "      <td>26489</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>3</td>\n",
       "      <td>25</td>\n",
       "      <td>MH8692_A</td>\n",
       "      <td>MH8692</td>\n",
       "      <td>A</td>\n",
       "      <td>41.106120</td>\n",
       "      <td>37.356550</td>\n",
       "      <td>64.933636</td>\n",
       "      <td>...</td>\n",
       "      <td>4.418474</td>\n",
       "      <td>1872.171013</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>132</td>\n",
       "      <td>21576.810909</td>\n",
       "      <td>32167.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26527</th>\n",
       "      <td>26528</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>4</td>\n",
       "      <td>69</td>\n",
       "      <td>MH8692_A</td>\n",
       "      <td>MH8692</td>\n",
       "      <td>A</td>\n",
       "      <td>41.704604</td>\n",
       "      <td>37.444830</td>\n",
       "      <td>63.727008</td>\n",
       "      <td>...</td>\n",
       "      <td>4.418474</td>\n",
       "      <td>1872.171013</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>132</td>\n",
       "      <td>21576.810909</td>\n",
       "      <td>31290.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26554</th>\n",
       "      <td>26555</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>7</td>\n",
       "      <td>85</td>\n",
       "      <td>MH8692_A</td>\n",
       "      <td>MH8692</td>\n",
       "      <td>A</td>\n",
       "      <td>41.063368</td>\n",
       "      <td>36.944469</td>\n",
       "      <td>60.461039</td>\n",
       "      <td>...</td>\n",
       "      <td>4.418474</td>\n",
       "      <td>1872.171013</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>132</td>\n",
       "      <td>21576.810909</td>\n",
       "      <td>25643.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26562</th>\n",
       "      <td>26563</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>10</td>\n",
       "      <td>113</td>\n",
       "      <td>MH8692_A</td>\n",
       "      <td>MH8692</td>\n",
       "      <td>A</td>\n",
       "      <td>41.463320</td>\n",
       "      <td>37.500386</td>\n",
       "      <td>63.997683</td>\n",
       "      <td>...</td>\n",
       "      <td>4.418474</td>\n",
       "      <td>1872.171013</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>132</td>\n",
       "      <td>21576.810909</td>\n",
       "      <td>31286.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26569</th>\n",
       "      <td>26570</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>11</td>\n",
       "      <td>119</td>\n",
       "      <td>MH8692_A</td>\n",
       "      <td>MH8692</td>\n",
       "      <td>A</td>\n",
       "      <td>41.003210</td>\n",
       "      <td>37.369181</td>\n",
       "      <td>63.711075</td>\n",
       "      <td>...</td>\n",
       "      <td>4.418474</td>\n",
       "      <td>1872.171013</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>132</td>\n",
       "      <td>21576.810909</td>\n",
       "      <td>27503.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26649</th>\n",
       "      <td>26650</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>MH8700_A</td>\n",
       "      <td>MH8700</td>\n",
       "      <td>A</td>\n",
       "      <td>41.106218</td>\n",
       "      <td>37.446891</td>\n",
       "      <td>63.032383</td>\n",
       "      <td>...</td>\n",
       "      <td>5.462005</td>\n",
       "      <td>20695.554297</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>133</td>\n",
       "      <td>22286.528748</td>\n",
       "      <td>31280.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26800</th>\n",
       "      <td>26801</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>MH8766</td>\n",
       "      <td>B</td>\n",
       "      <td>41.546301</td>\n",
       "      <td>37.384365</td>\n",
       "      <td>62.442997</td>\n",
       "      <td>...</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>136</td>\n",
       "      <td>18112.213595</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26801</th>\n",
       "      <td>26802</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>3</td>\n",
       "      <td>93</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>MH8766</td>\n",
       "      <td>B</td>\n",
       "      <td>41.447950</td>\n",
       "      <td>37.233438</td>\n",
       "      <td>63.975815</td>\n",
       "      <td>...</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>136</td>\n",
       "      <td>18112.213595</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26802</th>\n",
       "      <td>26803</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>4</td>\n",
       "      <td>120</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>MH8766</td>\n",
       "      <td>B</td>\n",
       "      <td>39.696233</td>\n",
       "      <td>37.439854</td>\n",
       "      <td>57.901580</td>\n",
       "      <td>...</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>136</td>\n",
       "      <td>18112.213595</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26803</th>\n",
       "      <td>26804</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>5</td>\n",
       "      <td>121</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>MH8766</td>\n",
       "      <td>B</td>\n",
       "      <td>39.934831</td>\n",
       "      <td>37.459551</td>\n",
       "      <td>57.384270</td>\n",
       "      <td>...</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>136</td>\n",
       "      <td>18112.213595</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26804</th>\n",
       "      <td>26805</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>6</td>\n",
       "      <td>170</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>MH8766</td>\n",
       "      <td>B</td>\n",
       "      <td>41.605440</td>\n",
       "      <td>37.343612</td>\n",
       "      <td>61.480751</td>\n",
       "      <td>...</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>136</td>\n",
       "      <td>18112.213595</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26805</th>\n",
       "      <td>26806</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>7</td>\n",
       "      <td>189</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>MH8766</td>\n",
       "      <td>B</td>\n",
       "      <td>41.413333</td>\n",
       "      <td>37.532549</td>\n",
       "      <td>59.916863</td>\n",
       "      <td>...</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>136</td>\n",
       "      <td>18112.213595</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26993</th>\n",
       "      <td>26994</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>13579</td>\n",
       "      <td>38</td>\n",
       "      <td>MH8766_B</td>\n",
       "      <td>MH8766</td>\n",
       "      <td>B</td>\n",
       "      <td>40.918681</td>\n",
       "      <td>37.307114</td>\n",
       "      <td>67.154772</td>\n",
       "      <td>...</td>\n",
       "      <td>6.371922</td>\n",
       "      <td>168187.646394</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>136</td>\n",
       "      <td>18112.213595</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27075</th>\n",
       "      <td>27076</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>3</td>\n",
       "      <td>163</td>\n",
       "      <td>MH8922_B</td>\n",
       "      <td>MH8922</td>\n",
       "      <td>B</td>\n",
       "      <td>41.704918</td>\n",
       "      <td>37.379098</td>\n",
       "      <td>63.010246</td>\n",
       "      <td>...</td>\n",
       "      <td>6.056122</td>\n",
       "      <td>81281.923343</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>205</td>\n",
       "      <td>17864.774079</td>\n",
       "      <td>19815.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27084</th>\n",
       "      <td>27085</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>1</td>\n",
       "      <td>152</td>\n",
       "      <td>MH8922_A</td>\n",
       "      <td>MH8922</td>\n",
       "      <td>A</td>\n",
       "      <td>40.993617</td>\n",
       "      <td>37.063830</td>\n",
       "      <td>65.325532</td>\n",
       "      <td>...</td>\n",
       "      <td>6.056122</td>\n",
       "      <td>81281.923343</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>137</td>\n",
       "      <td>18013.911137</td>\n",
       "      <td>17908.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27160</th>\n",
       "      <td>27161</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>0</td>\n",
       "      <td>57</td>\n",
       "      <td>MH8924_B</td>\n",
       "      <td>MH8924</td>\n",
       "      <td>B</td>\n",
       "      <td>41.671346</td>\n",
       "      <td>37.517754</td>\n",
       "      <td>64.987614</td>\n",
       "      <td>...</td>\n",
       "      <td>5.064007</td>\n",
       "      <td>8277.116060</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>206</td>\n",
       "      <td>16068.452506</td>\n",
       "      <td>20953.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27189</th>\n",
       "      <td>27190</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>1</td>\n",
       "      <td>90</td>\n",
       "      <td>MH8924_B</td>\n",
       "      <td>MH8924</td>\n",
       "      <td>B</td>\n",
       "      <td>41.802281</td>\n",
       "      <td>37.422053</td>\n",
       "      <td>63.293726</td>\n",
       "      <td>...</td>\n",
       "      <td>5.064007</td>\n",
       "      <td>8277.116060</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>206</td>\n",
       "      <td>16068.452506</td>\n",
       "      <td>17803.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27204</th>\n",
       "      <td>27205</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>MH8924_B</td>\n",
       "      <td>MH8924</td>\n",
       "      <td>B</td>\n",
       "      <td>41.526059</td>\n",
       "      <td>37.491857</td>\n",
       "      <td>54.752443</td>\n",
       "      <td>...</td>\n",
       "      <td>5.064007</td>\n",
       "      <td>8277.116060</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>206</td>\n",
       "      <td>16068.452506</td>\n",
       "      <td>11994.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27250</th>\n",
       "      <td>27251</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>3</td>\n",
       "      <td>39</td>\n",
       "      <td>MH8925_A</td>\n",
       "      <td>MH8925</td>\n",
       "      <td>A</td>\n",
       "      <td>41.588921</td>\n",
       "      <td>37.536443</td>\n",
       "      <td>61.093294</td>\n",
       "      <td>...</td>\n",
       "      <td>5.684998</td>\n",
       "      <td>34583.562088</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>139</td>\n",
       "      <td>19976.548232</td>\n",
       "      <td>32409.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27260</th>\n",
       "      <td>27261</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>4</td>\n",
       "      <td>52</td>\n",
       "      <td>MH8925_A</td>\n",
       "      <td>MH8925</td>\n",
       "      <td>A</td>\n",
       "      <td>41.602174</td>\n",
       "      <td>37.639130</td>\n",
       "      <td>61.995652</td>\n",
       "      <td>...</td>\n",
       "      <td>5.684998</td>\n",
       "      <td>34583.562088</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>139</td>\n",
       "      <td>19976.548232</td>\n",
       "      <td>16695.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27292</th>\n",
       "      <td>27293</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>7</td>\n",
       "      <td>122</td>\n",
       "      <td>MH8925_A</td>\n",
       "      <td>MH8925</td>\n",
       "      <td>A</td>\n",
       "      <td>40.516971</td>\n",
       "      <td>37.501305</td>\n",
       "      <td>75.872063</td>\n",
       "      <td>...</td>\n",
       "      <td>5.684998</td>\n",
       "      <td>34583.562088</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>139</td>\n",
       "      <td>19976.548232</td>\n",
       "      <td>8603.0</td>\n",
       "      <td>Synonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27310</th>\n",
       "      <td>27311</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>8</td>\n",
       "      <td>132</td>\n",
       "      <td>MH8925_A</td>\n",
       "      <td>MH8925</td>\n",
       "      <td>A</td>\n",
       "      <td>41.527886</td>\n",
       "      <td>37.487678</td>\n",
       "      <td>61.630350</td>\n",
       "      <td>...</td>\n",
       "      <td>5.684998</td>\n",
       "      <td>34583.562088</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>139</td>\n",
       "      <td>19976.548232</td>\n",
       "      <td>26549.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27335</th>\n",
       "      <td>27336</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>13558</td>\n",
       "      <td>8</td>\n",
       "      <td>MH9045_A</td>\n",
       "      <td>MH9045</td>\n",
       "      <td>A</td>\n",
       "      <td>41.616683</td>\n",
       "      <td>37.487918</td>\n",
       "      <td>63.541377</td>\n",
       "      <td>...</td>\n",
       "      <td>5.285284</td>\n",
       "      <td>13777.027742</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>141</td>\n",
       "      <td>20803.842571</td>\n",
       "      <td>37624.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27390</th>\n",
       "      <td>27391</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>2</td>\n",
       "      <td>106</td>\n",
       "      <td>MH9045_A</td>\n",
       "      <td>MH9045</td>\n",
       "      <td>A</td>\n",
       "      <td>41.651429</td>\n",
       "      <td>37.492381</td>\n",
       "      <td>62.947619</td>\n",
       "      <td>...</td>\n",
       "      <td>5.285284</td>\n",
       "      <td>13777.027742</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>141</td>\n",
       "      <td>20803.842571</td>\n",
       "      <td>28583.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27427</th>\n",
       "      <td>27428</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>7</td>\n",
       "      <td>229</td>\n",
       "      <td>MH9045_A</td>\n",
       "      <td>MH9045</td>\n",
       "      <td>A</td>\n",
       "      <td>41.553222</td>\n",
       "      <td>37.402607</td>\n",
       "      <td>60.401159</td>\n",
       "      <td>...</td>\n",
       "      <td>5.285284</td>\n",
       "      <td>13777.027742</td>\n",
       "      <td>N</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "      <td>141</td>\n",
       "      <td>20803.842571</td>\n",
       "      <td>23027.0</td>\n",
       "      <td>Nonsynonymous</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>579 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0       run   X1.x  Unnamed: 0.1        Id LAURING_ID  dup  \\\n",
       "0               1     perth      1           125      1102       1102  NaN   \n",
       "1               2     perth      3           216      1102       1102  NaN   \n",
       "274           275    cali09      1           108      1112       1112  NaN   \n",
       "327           328  cali09_2      1           173      1112       1112  NaN   \n",
       "422           423     perth      0            13      1121       1121  NaN   \n",
       "507           508     perth  13452           111      1121       1121  NaN   \n",
       "699           700   perth_2      1            95      1122       1122  NaN   \n",
       "876           877   perth_2      2            79      1125       1125  NaN   \n",
       "904           905   perth_2      3           103      1125       1125  NaN   \n",
       "1148         1149  cali09_2  13453            23      1137       1137  NaN   \n",
       "1152         1153  cali09_2  13461            42      1137       1137  NaN   \n",
       "1160         1161  cali09_2  13455            25      1137       1137  NaN   \n",
       "1164         1165  cali09_2  13448             5      1137       1137  NaN   \n",
       "1193         1194  cali09_2  13488           113      1137       1137  NaN   \n",
       "1197         1198  cali09_2  13490           116      1137       1137  NaN   \n",
       "1208         1209  cali09_2  13483           106      1137       1137  NaN   \n",
       "1212         1213  cali09_2  13489           115      1137       1137  NaN   \n",
       "1232         1233  cali09_2  13508           161      1137       1137  NaN   \n",
       "1236         1237  cali09_2  13511           169      1137       1137  NaN   \n",
       "1244         1245  cali09_2  13525           214      1137       1137  NaN   \n",
       "1247         1248  cali09_2  13532           228      1137       1137  NaN   \n",
       "1254         1255  cali09_2  13531           227      1137       1137  NaN   \n",
       "1260         1261  cali09_2  13526           215      1137       1137  NaN   \n",
       "1263         1264  cali09_2  13528           219      1137       1137  NaN   \n",
       "1266         1267  cali09_2  13535           235      1137       1137  NaN   \n",
       "1279         1280  cali09_2  13523           211      1137       1137  NaN   \n",
       "1283         1284  cali09_2  13554           282      1137       1137  NaN   \n",
       "1293         1294  cali09_2  13546           267      1137       1137  NaN   \n",
       "1297         1298  cali09_2  13557           292      1137       1137  NaN   \n",
       "1300         1301    cali09      2           248      1137       1137  NaN   \n",
       "...           ...       ...    ...           ...       ...        ...  ...   \n",
       "26147       26148      HK_1      6           120  MH8592_A     MH8592    A   \n",
       "26412       26413      HK_2  13581            70  MH8690_B     MH8690    B   \n",
       "26443       26444      HK_2      5           104  MH8690_B     MH8690    B   \n",
       "26478       26479      HK_1      1             6  MH8692_A     MH8692    A   \n",
       "26484       26485      HK_1      2            12  MH8692_A     MH8692    A   \n",
       "26488       26489      HK_1      3            25  MH8692_A     MH8692    A   \n",
       "26527       26528      HK_1      4            69  MH8692_A     MH8692    A   \n",
       "26554       26555      HK_1      7            85  MH8692_A     MH8692    A   \n",
       "26562       26563      HK_1     10           113  MH8692_A     MH8692    A   \n",
       "26569       26570      HK_1     11           119  MH8692_A     MH8692    A   \n",
       "26649       26650      HK_1      0            70  MH8700_A     MH8700    A   \n",
       "26800       26801      HK_1      1            19  MH8766_B     MH8766    B   \n",
       "26801       26802      HK_1      3            93  MH8766_B     MH8766    B   \n",
       "26802       26803      HK_1      4           120  MH8766_B     MH8766    B   \n",
       "26803       26804      HK_1      5           121  MH8766_B     MH8766    B   \n",
       "26804       26805      HK_1      6           170  MH8766_B     MH8766    B   \n",
       "26805       26806      HK_1      7           189  MH8766_B     MH8766    B   \n",
       "26993       26994      HK_1  13579            38  MH8766_B     MH8766    B   \n",
       "27075       27076      HK_2      3           163  MH8922_B     MH8922    B   \n",
       "27084       27085      HK_1      1           152  MH8922_A     MH8922    A   \n",
       "27160       27161      HK_2      0            57  MH8924_B     MH8924    B   \n",
       "27189       27190      HK_2      1            90  MH8924_B     MH8924    B   \n",
       "27204       27205      HK_2      2           100  MH8924_B     MH8924    B   \n",
       "27250       27251      HK_1      3            39  MH8925_A     MH8925    A   \n",
       "27260       27261      HK_1      4            52  MH8925_A     MH8925    A   \n",
       "27292       27293      HK_1      7           122  MH8925_A     MH8925    A   \n",
       "27310       27311      HK_1      8           132  MH8925_A     MH8925    A   \n",
       "27335       27336      HK_1  13558             8  MH9045_A     MH9045    A   \n",
       "27390       27391      HK_1      2           106  MH9045_A     MH9045    A   \n",
       "27427       27428      HK_1      7           229  MH9045_A     MH9045    A   \n",
       "\n",
       "            MapQ      Phred   Read_pos      ...       log_copy_num  \\\n",
       "0      41.672713  37.195615  63.517421      ...           6.486882   \n",
       "1      41.477381  36.948810  78.584524      ...           6.486882   \n",
       "274    41.383436  37.535106  61.826005      ...           6.044844   \n",
       "327    41.851569  37.230859  62.804341      ...           6.044844   \n",
       "422    38.713415  37.501220  62.269512      ...           5.060653   \n",
       "507    41.534214  37.391014  65.533185      ...           5.060653   \n",
       "699    41.909170  37.092561  63.629758      ...           4.315119   \n",
       "876    41.608527  37.136628  61.093346      ...           4.851537   \n",
       "904    41.486762  37.289206  63.089613      ...           4.851537   \n",
       "1148   41.472222  37.330409  61.217836      ...           4.711363   \n",
       "1152   40.915110  37.259762  63.876061      ...           4.711363   \n",
       "1160   41.772093  37.200000  64.624806      ...           4.711363   \n",
       "1164   41.438679  37.498428  60.407233      ...           4.711363   \n",
       "1193   41.549844  37.414330  65.526480      ...           4.711363   \n",
       "1197   41.793333  37.235000  61.713333      ...           4.711363   \n",
       "1208   41.820922  37.320922  62.187943      ...           4.711363   \n",
       "1212   41.679758  37.332326  61.658610      ...           4.711363   \n",
       "1232   41.734899  37.399329  63.546980      ...           4.711363   \n",
       "1236   41.740484  37.155709  61.806228      ...           4.711363   \n",
       "1244   41.594235  37.485588  61.647450      ...           4.711363   \n",
       "1247   41.147982  37.172646  63.387892      ...           4.711363   \n",
       "1254   41.549528  37.242925  64.320755      ...           4.711363   \n",
       "1260   41.537209  37.320930  61.581395      ...           4.711363   \n",
       "1263   41.860520  37.442080  61.669031      ...           4.711363   \n",
       "1266   40.905983  37.457265  64.395299      ...           4.711363   \n",
       "1279   41.588670  36.928571  63.371921      ...           4.711363   \n",
       "1283   41.795359  37.251055  58.856540      ...           4.711363   \n",
       "1293   41.602537  37.456660  59.617336      ...           4.711363   \n",
       "1297   41.448203  37.427061  65.701903      ...           4.711363   \n",
       "1300   38.980501  37.504178  60.540390      ...           4.711363   \n",
       "...          ...        ...        ...      ...                ...   \n",
       "26147  30.064327  37.286550  58.122807      ...           6.547221   \n",
       "26412  41.493766  37.441397  67.558603      ...           4.267601   \n",
       "26443  41.378205  37.435897  62.705128      ...           4.267601   \n",
       "26478  41.720973  37.457751  61.563526      ...           4.418474   \n",
       "26484  41.283700  37.460801  63.990976      ...           4.418474   \n",
       "26488  41.106120  37.356550  64.933636      ...           4.418474   \n",
       "26527  41.704604  37.444830  63.727008      ...           4.418474   \n",
       "26554  41.063368  36.944469  60.461039      ...           4.418474   \n",
       "26562  41.463320  37.500386  63.997683      ...           4.418474   \n",
       "26569  41.003210  37.369181  63.711075      ...           4.418474   \n",
       "26649  41.106218  37.446891  63.032383      ...           5.462005   \n",
       "26800  41.546301  37.384365  62.442997      ...           6.371922   \n",
       "26801  41.447950  37.233438  63.975815      ...           6.371922   \n",
       "26802  39.696233  37.439854  57.901580      ...           6.371922   \n",
       "26803  39.934831  37.459551  57.384270      ...           6.371922   \n",
       "26804  41.605440  37.343612  61.480751      ...           6.371922   \n",
       "26805  41.413333  37.532549  59.916863      ...           6.371922   \n",
       "26993  40.918681  37.307114  67.154772      ...           6.371922   \n",
       "27075  41.704918  37.379098  63.010246      ...           6.056122   \n",
       "27084  40.993617  37.063830  65.325532      ...           6.056122   \n",
       "27160  41.671346  37.517754  64.987614      ...           5.064007   \n",
       "27189  41.802281  37.422053  63.293726      ...           5.064007   \n",
       "27204  41.526059  37.491857  54.752443      ...           5.064007   \n",
       "27250  41.588921  37.536443  61.093294      ...           5.684998   \n",
       "27260  41.602174  37.639130  61.995652      ...           5.684998   \n",
       "27292  40.516971  37.501305  75.872063      ...           5.684998   \n",
       "27310  41.527886  37.487678  61.630350      ...           5.684998   \n",
       "27335  41.616683  37.487918  63.541377      ...           5.285284   \n",
       "27390  41.651429  37.492381  62.947619      ...           5.285284   \n",
       "27427  41.553222  37.402607  60.401159      ...           5.285284   \n",
       "\n",
       "               gc_ul  HIGHSD  sequenced  home_collected  snv_qualified   X1  \\\n",
       "0      219156.158420       N       True               0           True  403   \n",
       "1      219156.158420       N       True               0           True  403   \n",
       "274     79198.379568       N       True               0           True    1   \n",
       "327     79198.379568       N       True               0           True   54   \n",
       "422      8213.445551       N       True               0           True  409   \n",
       "507      8213.445551       N       True               0           True  409   \n",
       "699      1475.675381       N       True               0           True  465   \n",
       "876      5074.685473       N       True               0           True  466   \n",
       "904      5074.685473       N       True               0           True  466   \n",
       "1148     3674.806892       N       True               0           True   55   \n",
       "1152     3674.806892       N       True               0           True   55   \n",
       "1160     3674.806892       N       True               0           True   55   \n",
       "1164     3674.806892       N       True               0           True   55   \n",
       "1193     3674.806892       N       True               0           True   55   \n",
       "1197     3674.806892       N       True               0           True   55   \n",
       "1208     3674.806892       N       True               0           True   55   \n",
       "1212     3674.806892       N       True               0           True   55   \n",
       "1232     3674.806892       N       True               0           True   55   \n",
       "1236     3674.806892       N       True               0           True   55   \n",
       "1244     3674.806892       N       True               0           True   55   \n",
       "1247     3674.806892       N       True               0           True   55   \n",
       "1254     3674.806892       N       True               0           True   55   \n",
       "1260     3674.806892       N       True               0           True   55   \n",
       "1263     3674.806892       N       True               0           True   55   \n",
       "1266     3674.806892       N       True               0           True   55   \n",
       "1279     3674.806892       N       True               0           True   55   \n",
       "1283     3674.806892       N       True               0           True   55   \n",
       "1293     3674.806892       N       True               0           True   55   \n",
       "1297     3674.806892       N       True               0           True   55   \n",
       "1300     3674.806892       N       True               0           True    5   \n",
       "...              ...     ...        ...             ...            ...  ...   \n",
       "26147  251821.707282       N       True               0           True  125   \n",
       "26412    1322.734812       N       True               0           True  202   \n",
       "26443    1322.734812       N       True               0           True  202   \n",
       "26478    1872.171013       N       True               0           True  132   \n",
       "26484    1872.171013       N       True               0           True  132   \n",
       "26488    1872.171013       N       True               0           True  132   \n",
       "26527    1872.171013       N       True               0           True  132   \n",
       "26554    1872.171013       N       True               0           True  132   \n",
       "26562    1872.171013       N       True               0           True  132   \n",
       "26569    1872.171013       N       True               0           True  132   \n",
       "26649   20695.554297       N       True               0           True  133   \n",
       "26800  168187.646394       N       True               0           True  136   \n",
       "26801  168187.646394       N       True               0           True  136   \n",
       "26802  168187.646394       N       True               0           True  136   \n",
       "26803  168187.646394       N       True               0           True  136   \n",
       "26804  168187.646394       N       True               0           True  136   \n",
       "26805  168187.646394       N       True               0           True  136   \n",
       "26993  168187.646394       N       True               0           True  136   \n",
       "27075   81281.923343       N       True               0           True  205   \n",
       "27084   81281.923343       N       True               0           True  137   \n",
       "27160    8277.116060       N       True               0           True  206   \n",
       "27189    8277.116060       N       True               0           True  206   \n",
       "27204    8277.116060       N       True               0           True  206   \n",
       "27250   34583.562088       N       True               0           True  139   \n",
       "27260   34583.562088       N       True               0           True  139   \n",
       "27292   34583.562088       N       True               0           True  139   \n",
       "27310   34583.562088       N       True               0           True  139   \n",
       "27335   13777.027742       N       True               0           True  141   \n",
       "27390   13777.027742       N       True               0           True  141   \n",
       "27427   13777.027742       N       True               0           True  141   \n",
       "\n",
       "       sample_coverage  coverage   class_factor  \n",
       "0         29478.353879       NaN     Synonymous  \n",
       "1         29478.353879       NaN     Synonymous  \n",
       "274       17796.126556   27052.0  Nonsynonymous  \n",
       "327       16560.351942   14463.0     Synonymous  \n",
       "422       23610.190132   40503.0     Synonymous  \n",
       "507       23610.190132   22358.0     Synonymous  \n",
       "699       23418.068719   25880.0  Nonsynonymous  \n",
       "876       19125.357084   14718.0  Nonsynonymous  \n",
       "904       19125.357084   17276.0     Synonymous  \n",
       "1148      15518.761124   26321.0     Synonymous  \n",
       "1152      15518.761124   25227.0     Synonymous  \n",
       "1160      15518.761124   26539.0  Nonsynonymous  \n",
       "1164      15518.761124   23388.0     Synonymous  \n",
       "1193      15518.761124   23484.0     Synonymous  \n",
       "1197      15518.761124   22687.0     Synonymous  \n",
       "1208      15518.761124   22622.0     Synonymous  \n",
       "1212      15518.761124   23975.0     Synonymous  \n",
       "1232      15518.761124   14236.0     Synonymous  \n",
       "1236      15518.761124   13880.0     Synonymous  \n",
       "1244      15518.761124   16145.0     Synonymous  \n",
       "1247      15518.761124   15536.0     Synonymous  \n",
       "1254      15518.761124   17207.0     Synonymous  \n",
       "1260      15518.761124   16728.0  Nonsynonymous  \n",
       "1263      15518.761124   16941.0     Synonymous  \n",
       "1266      15518.761124   17076.0     Synonymous  \n",
       "1279      15518.761124   15210.0     Synonymous  \n",
       "1283      15518.761124   15795.0  Nonsynonymous  \n",
       "1293      15518.761124   14461.0     Synonymous  \n",
       "1297      15518.761124   17470.0     Synonymous  \n",
       "1300      12298.362302   11155.0  Nonsynonymous  \n",
       "...                ...       ...            ...  \n",
       "26147     23957.041996       NaN  Nonsynonymous  \n",
       "26412      3041.026127    3309.0     Synonymous  \n",
       "26443      3041.026127    3632.0     Synonymous  \n",
       "26478     21576.810909   43706.0  Nonsynonymous  \n",
       "26484     21576.810909   35421.0  Nonsynonymous  \n",
       "26488     21576.810909   32167.0  Nonsynonymous  \n",
       "26527     21576.810909   31290.0  Nonsynonymous  \n",
       "26554     21576.810909   25643.0     Synonymous  \n",
       "26562     21576.810909   31286.0  Nonsynonymous  \n",
       "26569     21576.810909   27503.0     Synonymous  \n",
       "26649     22286.528748   31280.0     Synonymous  \n",
       "26800     18112.213595       NaN     Synonymous  \n",
       "26801     18112.213595       NaN  Nonsynonymous  \n",
       "26802     18112.213595       NaN     Synonymous  \n",
       "26803     18112.213595       NaN     Synonymous  \n",
       "26804     18112.213595       NaN     Synonymous  \n",
       "26805     18112.213595       NaN     Synonymous  \n",
       "26993     18112.213595       NaN  Nonsynonymous  \n",
       "27075     17864.774079   19815.0     Synonymous  \n",
       "27084     18013.911137   17908.0     Synonymous  \n",
       "27160     16068.452506   20953.0  Nonsynonymous  \n",
       "27189     16068.452506   17803.0     Synonymous  \n",
       "27204     16068.452506   11994.0  Nonsynonymous  \n",
       "27250     19976.548232   32409.0     Synonymous  \n",
       "27260     19976.548232   16695.0  Nonsynonymous  \n",
       "27292     19976.548232    8603.0     Synonymous  \n",
       "27310     19976.548232   26549.0  Nonsynonymous  \n",
       "27335     20803.842571   37624.0  Nonsynonymous  \n",
       "27390     20803.842571   28583.0  Nonsynonymous  \n",
       "27427     20803.842571   23027.0  Nonsynonymous  \n",
       "\n",
       "[579 rows x 54 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is a function that locates each HA mutation in the PDB structures. This is the equivalent of H3 numbering for H3 strains. (Assigns codon 1 to codon 17 or (AA pos - 16))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def locate(snv,run_label,ref_fa,antigenic):\n",
    "    # Get the HA seqeunce the sample was aligned to - the cordinates here match those in the dataframe\n",
    "    for seq_record in SeqIO.parse(ref_fa, \"fasta\"):    \n",
    "        if seq_record.name == \"HA\":\n",
    "            HA=seq_record\n",
    "            HA_prot = seq_record.seq.translate(to_stop=True) # translate to protien.\n",
    "    run_snv = snv.loc[(snv[\"run\"].str.contains(run_label))  & (snv[\"chr\"]==\"HA\")] # Get the isnv from that run and that in HA \n",
    "    PDB_4HMG=[]\n",
    "    PDB_4JTV=[]\n",
    "    for index, var in run_snv.iterrows():\n",
    "        nice = find_aligments('muscle',\"/Users/jt/muscle3.8.31/\",HA_prot,var[\"AA_pos\"]) # These positions are base 1 (see below) whcih is what the function expects\n",
    "        PDB_4HMG.append(nice[0][0])\n",
    "        PDB_4JTV.append(nice[1][0])\n",
    "    run_snv[\"PDB_4HMG\"] = PDB_4HMG\n",
    "    run_snv[\"PDB_4JTV\"] = PDB_4JTV\n",
    "    return(run_snv)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Perth 2010-2011 & 2011-2012"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue H172 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K156 in HA1 in 4HMG\n",
      "  * K159 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D101 in sequential numbering of your HA sequence corresponds to:\n",
      "  * D85 in HA1 in 4HMG\n",
      "  * S83 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L102 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L86 in HA1 in 4HMG\n",
      "  * Y84 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E78 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I62 in HA1 in 4HMG\n",
      "  * G59 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A554 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T3 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I10 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue N24 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N8 in HA1 in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K557 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue V313 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V297 in HA1 in 4HMG\n",
      "  * I301 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E105 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E89 in HA1 in 4HMG\n",
      "  * E87 of HA1 in 4JTV\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "perth = locate(snv,\"perth\",\"/Users/jt/Documents/Analysis/HIVE/data/processed/perth/coding_fa/Perth_mp.removed.parsed.coding.fasta\",antigenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Victoria 2012-2013 & 2013-2014"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G197 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G181 in HA1 in 4HMG\n",
      "  * G184 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I230 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I214 in HA1 in 4HMG\n",
      "  * K217 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K524 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T133 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T117 in HA1 in 4HMG\n",
      "  * R119 of HA1 in 4JTV\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "victoria = locate(snv,\"vic\",\"/Users/jt/Documents/Analysis/HIVE/data/processed/victoria/coding_fa/Vic_pool.removed.parsed.coding.fasta\",antigenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hong Kong 2014-2015"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P322 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P306 in HA1 in 4HMG\n",
      "  * P310 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D307 in sequential numbering of your HA sequence corresponds to:\n",
      "  * D291 in HA1 in 4HMG\n",
      "  * S295 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I258 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V242 in HA1 in 4HMG\n",
      "  * K245 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A122 in sequential numbering of your HA sequence corresponds to:\n",
      "  * A106 in HA1 in 4HMG\n",
      "  * E105 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P37 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P21 in HA1 in 4HMG\n",
      "  * N17 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L547 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D69 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N53 in HA1 in 4HMG\n",
      "  * K49 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S328 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N312 in HA1 in 4HMG\n",
      "  * T316 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue Y486 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Y141 in HA2 in 4HMG\n",
      "  * Y141 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S165 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S149 in HA1 in 4HMG\n",
      "  * K152 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T203 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T187 in HA1 in 4HMG\n",
      "  * T190 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue M284 in sequential numbering of your HA sequence corresponds to:\n",
      "  * M268 in HA1 in 4HMG\n",
      "  * I272 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue H172 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K156 in HA1 in 4HMG\n",
      "  * K159 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L82 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L66 in HA1 in 4HMG\n",
      "  * I63 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I258 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V242 in HA1 in 4HMG\n",
      "  * K245 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K205 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q189 in HA1 in 4HMG\n",
      "  * A192 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue H451 in sequential numbering of your HA sequence corresponds to:\n",
      "  * H106 in HA2 in 4HMG\n",
      "  * R106 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue F348 in sequential numbering of your HA sequence corresponds to:\n",
      "  * F3 in HA2 in 4HMG\n",
      "  * F3 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L330 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L314 in HA1 in 4HMG\n",
      "  * L318 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A154 in sequential numbering of your HA sequence corresponds to:\n",
      "  * A138 in HA1 in 4HMG\n",
      "  * A141 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S130 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S114 in HA1 in 4HMG\n",
      "  * S116 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R323 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K307 in HA1 in 4HMG\n",
      "  * K311 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P322 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P306 in HA1 in 4HMG\n",
      "  * P310 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E465 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E120 in HA2 in 4HMG\n",
      "  * E120 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A212 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V196 in HA1 in 4HMG\n",
      "  * Q199 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue M478 in sequential numbering of your HA sequence corresponds to:\n",
      "  * M133 in HA2 in 4HMG\n",
      "  * I133 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P37 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P21 in HA1 in 4HMG\n",
      "  * N17 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L547 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L332 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L316 in HA1 in 4HMG\n",
      "  * L320 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A179 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V163 in HA1 in 4HMG\n",
      "  * K166 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G495 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E150 in HA2 in 4HMG\n",
      "  * E150 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P340 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P324 in HA1 in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue V339 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V323 in HA1 in 4HMG\n",
      "  * I327 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue F209 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S193 in HA1 in 4HMG\n",
      "  * S196 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S165 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S149 in HA1 in 4HMG\n",
      "  * K152 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T203 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T187 in HA1 in 4HMG\n",
      "  * T190 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T144 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T128 in HA1 in 4HMG\n",
      "  * P130 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue H172 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K156 in HA1 in 4HMG\n",
      "  * K159 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S278 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T262 in HA1 in 4HMG\n",
      "  * N266 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L383 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L38 in HA2 in 4HMG\n",
      "  * L38 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue C482 in sequential numbering of your HA sequence corresponds to:\n",
      "  * C137 in HA2 in 4HMG\n",
      "  * C137 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K468 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R123 in HA2 in 4HMG\n",
      "  * R123 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue F415 in sequential numbering of your HA sequence corresponds to:\n",
      "  * F70 in HA2 in 4HMG\n",
      "  * F70 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue N499 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N154 in HA2 in 4HMG\n",
      "  * N154 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D101 in sequential numbering of your HA sequence corresponds to:\n",
      "  * D85 in HA1 in 4HMG\n",
      "  * S83 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue A288 in sequential numbering of your HA sequence corresponds to:\n",
      "  * A272 in HA1 in 4HMG\n",
      "  * T276 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G279 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G263 in HA1 in 4HMG\n",
      "  * A267 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R323 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K307 in HA1 in 4HMG\n",
      "  * K311 of HA1 in 4JTV\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "HK = locate(snv,\"HK\",\"/Users/jt/Documents/Analysis/HIVE/data/processed/HK_6/coding_fa/PC1A.removed.parsed.coding.fasta\",antigenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cali09 H1N1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E241 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S227 in HA1 in 4HMG\n",
      "  * E230 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L442 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L98 in HA2 in 4HMG\n",
      "  * L98 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T262 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N248 in HA1 in 4HMG\n",
      "  * T251 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R130 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T117 in HA1 in 4HMG\n",
      "  * R119 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue E132 in sequential numbering of your HA sequence corresponds to:\n",
      "  * E119 in HA1 in 4HMG\n",
      "  * E121 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue P199 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P185 in HA1 in 4HMG\n",
      "  * P188 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S145 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q132 in HA1 in 4HMG\n",
      "  * S134 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R222 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R208 in HA1 in 4HMG\n",
      "  * R211 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue N404 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N60 in HA2 in 4HMG\n",
      "  * N60 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G172 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G158 in HA1 in 4HMG\n",
      "  * G161 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue F432 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K88 in HA2 in 4HMG\n",
      "  * F88 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D489 in sequential numbering of your HA sequence corresponds to:\n",
      "  * D145 in HA2 in 4HMG\n",
      "  * D145 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue G23 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G16 in HA1 in 4HMG\n",
      "  * G12 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K511 in sequential numbering of your HA sequence corresponds to:\n",
      "  * L167 in HA2 in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue V149 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * V138 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue Q210 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V196 in HA1 in 4HMG\n",
      "  * Q199 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making MUSCLE alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T451 in sequential numbering of your HA sequence corresponds to:\n",
      "  * T107 in HA2 in 4HMG\n",
      "  * T107 of HA2 in 4JTV\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:15: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "cali09 = locate(snv,\"cali\",\"/Users/jt/Documents/Analysis/HIVE/data/processed/cali09/coding_fa/Cali_pool.removed.parsed.coding.fasta\",antigenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# All sites\n",
    "\n",
    "Here we combine all the runs together, select for the nonsynonymous mutations and make a subtype column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  app.launch_new_instance()\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/pandas/core/indexing.py:465: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  self.obj[item] = s\n"
     ]
    }
   ],
   "source": [
    "HA_All = perth.append([victoria,HK,cali09], ignore_index = True)\n",
    "HA_nonsyn = HA_All.loc[HA_All[\"class_factor\"]==\"Nonsynonymous\"]\n",
    "HA_nonsyn[\"subtype\"] = \"H3\"\n",
    "HA_nonsyn.loc[HA_nonsyn[\"run\"]==\"cali09\",\"subtype\"]=\"H1\"\n",
    "HA_nonsyn.loc[HA_nonsyn[\"run\"]==\"cali09_2\",\"subtype\"]=\"H1\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will us regex to split the PDB names into amino acid position and segement (HA1- HA2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "loc=[]\n",
    "seg = []\n",
    "pat = re.compile('\\w(\\d+) in (HA[1,2])') \n",
    "for index, row in HA_nonsyn.iterrows():\n",
    "    s = pat.match(row.loc[\"PDB_4HMG\"])\n",
    "    if s:\n",
    "        loc.append(s.group(1))\n",
    "        seg.append(s.group(2))\n",
    "    else:\n",
    "        loc.append(\"-\")\n",
    "        seg.append(\"-\")\n",
    "HA_nonsyn[\"H3_pos\"]=loc\n",
    "HA_nonsyn[\"H3_seg\"]=seg"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, which of these positions are in antigenic sites?\n",
    "\n",
    "For this work I am cycling through each row of the data frame. \n",
    "    \n",
    "If the row is for an H3 virus in with a mutation in the HA1 segment Then we find the antigenic site in the frist try statement. I then assign the source in the second try statement.\n",
    "    \n",
    "    \n",
    "If the row is in an H1 virus then we check if the PBD_4JTV position is in the PBD_4JTV list of antigenic sites. All these came from Caton."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:28: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "antigenic_sites=[]\n",
    "source = []\n",
    "for index, row in HA_nonsyn.iterrows():\n",
    "    if row.loc[\"subtype\"]==\"H3\" and row.loc[\"H3_seg\"]==\"HA1\":\n",
    "        try:\n",
    "            antigenic_sites.append(H3.loc[H3['aa']==int(row.loc[\"H3_pos\"]),\"Antigenic sites\"].values[0])\n",
    "        except IndexError:\n",
    "            antigenic_sites.append(\"NA\")\n",
    "        try:\n",
    "            s = H3.loc[H3['aa']==int(row.loc[\"H3_pos\"]),\"Source\"].values[0]\n",
    "            if pd.isnull(s): # If there is no input then the site was mentioned in Lee et al. 2004\n",
    "                s=\"Lee\"\n",
    "            source.append(s)\n",
    "        except IndexError:\n",
    "            source.append(\"NA\")\n",
    "        \n",
    "    elif row.loc[\"subtype\"]==\"H1\":\n",
    "        if row.loc[\"PDB_4JTV\"] in antigenic:\n",
    "            antigenic_sites.append(\"H1N1\")\n",
    "            source.append(\"Caton\")\n",
    "        else:\n",
    "            antigenic_sites.append(\"NA\")\n",
    "            source.append(\"NA\")\n",
    "    else:\n",
    "        antigenic_sites.append(\"NA\")\n",
    "        source.append(\"NA\")\n",
    "HA_nonsyn[\"Antigenic\"] = antigenic_sites\n",
    "HA_nonsyn[\"Source\"] = source"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here are the nonsynonymous mutations in HA with the antigenic information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SPECID</th>\n",
       "      <th>onset</th>\n",
       "      <th>mutation</th>\n",
       "      <th>freq.var</th>\n",
       "      <th>Ref_AA</th>\n",
       "      <th>AA_pos</th>\n",
       "      <th>Var_AA</th>\n",
       "      <th>PDB_4JTV</th>\n",
       "      <th>PDB_4HMG</th>\n",
       "      <th>run</th>\n",
       "      <th>Antigenic</th>\n",
       "      <th>Source</th>\n",
       "      <th>subtype</th>\n",
       "      <th>H3_pos</th>\n",
       "      <th>H3_seg</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M54724</td>\n",
       "      <td>2011-03-20</td>\n",
       "      <td>HA_C329A</td>\n",
       "      <td>0.087608</td>\n",
       "      <td>['L']</td>\n",
       "      <td>[102]</td>\n",
       "      <td>['I']</td>\n",
       "      <td>Y84 of HA1</td>\n",
       "      <td>L86 in HA1</td>\n",
       "      <td>perth</td>\n",
       "      <td>E</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>86</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M54913</td>\n",
       "      <td>2011-03-30</td>\n",
       "      <td>HA_A258G</td>\n",
       "      <td>0.071215</td>\n",
       "      <td>['E']</td>\n",
       "      <td>[78]</td>\n",
       "      <td>['G']</td>\n",
       "      <td>G59 of HA1</td>\n",
       "      <td>I62 in HA1</td>\n",
       "      <td>perth_2</td>\n",
       "      <td>E</td>\n",
       "      <td>Smith</td>\n",
       "      <td>H3</td>\n",
       "      <td>62</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>M55001</td>\n",
       "      <td>2011-04-01</td>\n",
       "      <td>HA_T54T</td>\n",
       "      <td>0.032427</td>\n",
       "      <td>['T']</td>\n",
       "      <td>[10]</td>\n",
       "      <td>['I']</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>perth</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MH0778</td>\n",
       "      <td>2012-02-13</td>\n",
       "      <td>HA_T963C</td>\n",
       "      <td>0.071359</td>\n",
       "      <td>['V']</td>\n",
       "      <td>[313]</td>\n",
       "      <td>['A']</td>\n",
       "      <td>I301 of HA1</td>\n",
       "      <td>V297 in HA1</td>\n",
       "      <td>perth_2</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>297</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>MH2436</td>\n",
       "      <td>2012-12-10</td>\n",
       "      <td>HA_T718C</td>\n",
       "      <td>0.065651</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[230]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>K217 of HA1</td>\n",
       "      <td>I214 in HA1</td>\n",
       "      <td>victoria</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>214</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>MH4091</td>\n",
       "      <td>2013-03-05</td>\n",
       "      <td>HA_A1599G</td>\n",
       "      <td>0.079685</td>\n",
       "      <td>['K']</td>\n",
       "      <td>[524]</td>\n",
       "      <td>['E']</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>victoria</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HS1284</td>\n",
       "      <td>2014-11-19</td>\n",
       "      <td>HA_G948A</td>\n",
       "      <td>0.048948</td>\n",
       "      <td>['D']</td>\n",
       "      <td>[307]</td>\n",
       "      <td>['N']</td>\n",
       "      <td>S295 of HA1</td>\n",
       "      <td>D291 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>291</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HS1294</td>\n",
       "      <td>2014-11-18</td>\n",
       "      <td>HA_A801T</td>\n",
       "      <td>0.163603</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[258]</td>\n",
       "      <td>['L']</td>\n",
       "      <td>K245 of HA1</td>\n",
       "      <td>V242 in HA1</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>242</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HS1294</td>\n",
       "      <td>2014-11-18</td>\n",
       "      <td>HA_G393A</td>\n",
       "      <td>0.066540</td>\n",
       "      <td>['A']</td>\n",
       "      <td>[122]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>E105 of HA1</td>\n",
       "      <td>A106 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>106</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HS1376</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_C236A</td>\n",
       "      <td>0.025191</td>\n",
       "      <td>['D']</td>\n",
       "      <td>[69]</td>\n",
       "      <td>['E']</td>\n",
       "      <td>K49 of HA1</td>\n",
       "      <td>N53 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>C</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>H3</td>\n",
       "      <td>53</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HS1376</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_A1011G</td>\n",
       "      <td>0.023458</td>\n",
       "      <td>['S']</td>\n",
       "      <td>[328]</td>\n",
       "      <td>['G']</td>\n",
       "      <td>T316 of HA1</td>\n",
       "      <td>N312 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>312</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HS1381</td>\n",
       "      <td>2014-12-05</td>\n",
       "      <td>HA_A1486T</td>\n",
       "      <td>0.103901</td>\n",
       "      <td>['Y']</td>\n",
       "      <td>[486]</td>\n",
       "      <td>['F']</td>\n",
       "      <td>Y141 of HA2</td>\n",
       "      <td>Y141 in HA2</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>141</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>HS1401</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_A879C</td>\n",
       "      <td>0.034116</td>\n",
       "      <td>['M']</td>\n",
       "      <td>[284]</td>\n",
       "      <td>['L']</td>\n",
       "      <td>I272 of HA1</td>\n",
       "      <td>M268 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>268</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>HS1425</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_T802C</td>\n",
       "      <td>0.028037</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[258]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>K245 of HA1</td>\n",
       "      <td>V242 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>242</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>HS1450</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>HA_G644T</td>\n",
       "      <td>0.037827</td>\n",
       "      <td>['K']</td>\n",
       "      <td>[205]</td>\n",
       "      <td>['N']</td>\n",
       "      <td>A192 of HA1</td>\n",
       "      <td>Q189 in HA1</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>B</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>H3</td>\n",
       "      <td>189</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>HS1455</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>HA_A1381A</td>\n",
       "      <td>0.350988</td>\n",
       "      <td>['R']</td>\n",
       "      <td>[451]</td>\n",
       "      <td>['H']</td>\n",
       "      <td>R106 of HA2</td>\n",
       "      <td>H106 in HA2</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>106</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>HS1473</td>\n",
       "      <td>2014-12-15</td>\n",
       "      <td>HA_T1071C</td>\n",
       "      <td>0.177116</td>\n",
       "      <td>['F']</td>\n",
       "      <td>[348]</td>\n",
       "      <td>['L']</td>\n",
       "      <td>F3 of HA2</td>\n",
       "      <td>F3 in HA2</td>\n",
       "      <td>HK_6</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>3</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>HS1595</td>\n",
       "      <td>2015-01-04</td>\n",
       "      <td>HA_C418T</td>\n",
       "      <td>0.031555</td>\n",
       "      <td>['S']</td>\n",
       "      <td>[130]</td>\n",
       "      <td>['L']</td>\n",
       "      <td>S116 of HA1</td>\n",
       "      <td>S114 in HA1</td>\n",
       "      <td>HK_7</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>114</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>HS1604</td>\n",
       "      <td>2015-01-11</td>\n",
       "      <td>HA_G997G</td>\n",
       "      <td>0.118966</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[323]</td>\n",
       "      <td>['R']</td>\n",
       "      <td>K311 of HA1</td>\n",
       "      <td>K307 in HA1</td>\n",
       "      <td>HK_7</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>307</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>MH7611</td>\n",
       "      <td>2014-11-23</td>\n",
       "      <td>HA_G1463A</td>\n",
       "      <td>0.043976</td>\n",
       "      <td>['M']</td>\n",
       "      <td>[478]</td>\n",
       "      <td>['I']</td>\n",
       "      <td>I133 of HA2</td>\n",
       "      <td>M133 in HA2</td>\n",
       "      <td>HK_7</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>133</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>MH7755</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_G564A</td>\n",
       "      <td>0.032187</td>\n",
       "      <td>['A']</td>\n",
       "      <td>[179]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>K166 of HA1</td>\n",
       "      <td>V163 in HA1</td>\n",
       "      <td>HK_7</td>\n",
       "      <td>B</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>163</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>MH7796</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_G1512A</td>\n",
       "      <td>0.111589</td>\n",
       "      <td>['G']</td>\n",
       "      <td>[495]</td>\n",
       "      <td>['R']</td>\n",
       "      <td>E150 of HA2</td>\n",
       "      <td>E150 in HA2</td>\n",
       "      <td>HK_7</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>150</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>MH7851</td>\n",
       "      <td>2014-12-05</td>\n",
       "      <td>HA_T655C</td>\n",
       "      <td>0.174563</td>\n",
       "      <td>['F']</td>\n",
       "      <td>[209]</td>\n",
       "      <td>['S']</td>\n",
       "      <td>S196 of HA1</td>\n",
       "      <td>S193 in HA1</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>B</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>H3</td>\n",
       "      <td>193</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>MH7860</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_A459G</td>\n",
       "      <td>0.110522</td>\n",
       "      <td>['T']</td>\n",
       "      <td>[144]</td>\n",
       "      <td>['A']</td>\n",
       "      <td>P130 of HA1</td>\n",
       "      <td>T128 in HA1</td>\n",
       "      <td>HK_8</td>\n",
       "      <td>B</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>128</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>MH8052</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>HA_G862A</td>\n",
       "      <td>0.030123</td>\n",
       "      <td>['S']</td>\n",
       "      <td>[278]</td>\n",
       "      <td>['N']</td>\n",
       "      <td>N266 of HA1</td>\n",
       "      <td>T262 in HA1</td>\n",
       "      <td>HK_8</td>\n",
       "      <td>E</td>\n",
       "      <td>Smith</td>\n",
       "      <td>H3</td>\n",
       "      <td>262</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>MH8308</td>\n",
       "      <td>2014-12-16</td>\n",
       "      <td>HA_T1526A</td>\n",
       "      <td>0.041457</td>\n",
       "      <td>['N']</td>\n",
       "      <td>[499]</td>\n",
       "      <td>['K']</td>\n",
       "      <td>N154 of HA2</td>\n",
       "      <td>N154 in HA2</td>\n",
       "      <td>HK_2</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>154</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>MH8692</td>\n",
       "      <td>2014-12-27</td>\n",
       "      <td>HA_G891T</td>\n",
       "      <td>0.074946</td>\n",
       "      <td>['A']</td>\n",
       "      <td>[288]</td>\n",
       "      <td>['S']</td>\n",
       "      <td>T276 of HA1</td>\n",
       "      <td>A272 in HA1</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H3</td>\n",
       "      <td>272</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>MH9045</td>\n",
       "      <td>2015-01-11</td>\n",
       "      <td>HA_G997G</td>\n",
       "      <td>0.160589</td>\n",
       "      <td>['I']</td>\n",
       "      <td>[323]</td>\n",
       "      <td>['R']</td>\n",
       "      <td>K311 of HA1</td>\n",
       "      <td>K307 in HA1</td>\n",
       "      <td>HK_1</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>H3</td>\n",
       "      <td>307</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>M54062</td>\n",
       "      <td>2011-02-13</td>\n",
       "      <td>HA_C805C</td>\n",
       "      <td>0.024304</td>\n",
       "      <td>['N']</td>\n",
       "      <td>[262]</td>\n",
       "      <td>['T']</td>\n",
       "      <td>T251 of HA1</td>\n",
       "      <td>N248 in HA1</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H1</td>\n",
       "      <td>248</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>MH2936</td>\n",
       "      <td>2013-01-12</td>\n",
       "      <td>HA_C616A</td>\n",
       "      <td>0.024214</td>\n",
       "      <td>['P']</td>\n",
       "      <td>[199]</td>\n",
       "      <td>['Q']</td>\n",
       "      <td>P188 of HA1</td>\n",
       "      <td>P185 in HA1</td>\n",
       "      <td>cali09</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H1</td>\n",
       "      <td>185</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>MH5405</td>\n",
       "      <td>2013-12-14</td>\n",
       "      <td>HA_G685A</td>\n",
       "      <td>0.020723</td>\n",
       "      <td>['R']</td>\n",
       "      <td>[222]</td>\n",
       "      <td>['K']</td>\n",
       "      <td>R211 of HA1</td>\n",
       "      <td>R208 in HA1</td>\n",
       "      <td>cali09</td>\n",
       "      <td>H1N1</td>\n",
       "      <td>Caton</td>\n",
       "      <td>H1</td>\n",
       "      <td>208</td>\n",
       "      <td>HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>MH5546</td>\n",
       "      <td>2013-12-23</td>\n",
       "      <td>HA_G1485A</td>\n",
       "      <td>0.032572</td>\n",
       "      <td>['D']</td>\n",
       "      <td>[489]</td>\n",
       "      <td>['N']</td>\n",
       "      <td>D145 of HA2</td>\n",
       "      <td>D145 in HA2</td>\n",
       "      <td>cali09_2</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H1</td>\n",
       "      <td>145</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>MH5817</td>\n",
       "      <td>2014-01-10</td>\n",
       "      <td>HA_C1372A</td>\n",
       "      <td>0.024559</td>\n",
       "      <td>['T']</td>\n",
       "      <td>[451]</td>\n",
       "      <td>['N']</td>\n",
       "      <td>T107 of HA2</td>\n",
       "      <td>T107 in HA2</td>\n",
       "      <td>cali09</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "      <td>H1</td>\n",
       "      <td>107</td>\n",
       "      <td>HA2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    SPECID       onset   mutation  freq.var Ref_AA AA_pos Var_AA  \\\n",
       "2   M54724  2011-03-20   HA_C329A  0.087608  ['L']  [102]  ['I']   \n",
       "3   M54913  2011-03-30   HA_A258G  0.071215  ['E']   [78]  ['G']   \n",
       "6   M55001  2011-04-01    HA_T54T  0.032427  ['T']   [10]  ['I']   \n",
       "9   MH0778  2012-02-13   HA_T963C  0.071359  ['V']  [313]  ['A']   \n",
       "12  MH2436  2012-12-10   HA_T718C  0.065651  ['I']  [230]  ['T']   \n",
       "13  MH4091  2013-03-05  HA_A1599G  0.079685  ['K']  [524]  ['E']   \n",
       "16  HS1284  2014-11-19   HA_G948A  0.048948  ['D']  [307]  ['N']   \n",
       "17  HS1294  2014-11-18   HA_A801T  0.163603  ['I']  [258]  ['L']   \n",
       "18  HS1294  2014-11-18   HA_G393A  0.066540  ['A']  [122]  ['T']   \n",
       "21  HS1376  2014-12-03   HA_C236A  0.025191  ['D']   [69]  ['E']   \n",
       "22  HS1376  2014-12-03  HA_A1011G  0.023458  ['S']  [328]  ['G']   \n",
       "23  HS1381  2014-12-05  HA_A1486T  0.103901  ['Y']  [486]  ['F']   \n",
       "26  HS1401  2014-12-06   HA_A879C  0.034116  ['M']  [284]  ['L']   \n",
       "29  HS1425  2014-12-06   HA_T802C  0.028037  ['I']  [258]  ['T']   \n",
       "30  HS1450  2014-12-13   HA_G644T  0.037827  ['K']  [205]  ['N']   \n",
       "31  HS1455  2014-12-13  HA_A1381A  0.350988  ['R']  [451]  ['H']   \n",
       "32  HS1473  2014-12-15  HA_T1071C  0.177116  ['F']  [348]  ['L']   \n",
       "35  HS1595  2015-01-04   HA_C418T  0.031555  ['S']  [130]  ['L']   \n",
       "36  HS1604  2015-01-11   HA_G997G  0.118966  ['I']  [323]  ['R']   \n",
       "40  MH7611  2014-11-23  HA_G1463A  0.043976  ['M']  [478]  ['I']   \n",
       "44  MH7755  2014-12-03   HA_G564A  0.032187  ['A']  [179]  ['T']   \n",
       "45  MH7796  2014-12-06  HA_G1512A  0.111589  ['G']  [495]  ['R']   \n",
       "48  MH7851  2014-12-05   HA_T655C  0.174563  ['F']  [209]  ['S']   \n",
       "51  MH7860  2014-12-06   HA_A459G  0.110522  ['T']  [144]  ['A']   \n",
       "53  MH8052  2014-12-13   HA_G862A  0.030123  ['S']  [278]  ['N']   \n",
       "58  MH8308  2014-12-16  HA_T1526A  0.041457  ['N']  [499]  ['K']   \n",
       "60  MH8692  2014-12-27   HA_G891T  0.074946  ['A']  [288]  ['S']   \n",
       "62  MH9045  2015-01-11   HA_G997G  0.160589  ['I']  [323]  ['R']   \n",
       "65  M54062  2011-02-13   HA_C805C  0.024304  ['N']  [262]  ['T']   \n",
       "68  MH2936  2013-01-12   HA_C616A  0.024214  ['P']  [199]  ['Q']   \n",
       "70  MH5405  2013-12-14   HA_G685A  0.020723  ['R']  [222]  ['K']   \n",
       "74  MH5546  2013-12-23  HA_G1485A  0.032572  ['D']  [489]  ['N']   \n",
       "79  MH5817  2014-01-10  HA_C1372A  0.024559  ['T']  [451]  ['N']   \n",
       "\n",
       "             PDB_4JTV           PDB_4HMG       run Antigenic Source subtype  \\\n",
       "2         Y84 of HA1         L86 in HA1      perth         E    Lee      H3   \n",
       "3         G59 of HA1         I62 in HA1    perth_2         E  Smith      H3   \n",
       "6   an alignment gap   an alignment gap      perth        NA     NA      H3   \n",
       "9        I301 of HA1        V297 in HA1    perth_2         C    Lee      H3   \n",
       "12       K217 of HA1        I214 in HA1   victoria         D    Lee      H3   \n",
       "13  an alignment gap   an alignment gap   victoria        NA     NA      H3   \n",
       "16       S295 of HA1        D291 in HA1       HK_6        NA     NA      H3   \n",
       "17       K245 of HA1        V242 in HA1       HK_1         D    Lee      H3   \n",
       "18       E105 of HA1        A106 in HA1       HK_6        NA     NA      H3   \n",
       "21        K49 of HA1         N53 in HA1       HK_6         C  Wiley      H3   \n",
       "22       T316 of HA1        N312 in HA1       HK_6         C    Lee      H3   \n",
       "23       Y141 of HA2        Y141 in HA2       HK_6        NA     NA      H3   \n",
       "26       I272 of HA1        M268 in HA1       HK_6        NA     NA      H3   \n",
       "29       K245 of HA1        V242 in HA1       HK_6         D    Lee      H3   \n",
       "30       A192 of HA1        Q189 in HA1       HK_6         B  Wiley      H3   \n",
       "31       R106 of HA2        H106 in HA2       HK_1        NA     NA      H3   \n",
       "32         F3 of HA2          F3 in HA2       HK_6        NA     NA      H3   \n",
       "35       S116 of HA1        S114 in HA1       HK_7        NA     NA      H3   \n",
       "36       K311 of HA1        K307 in HA1       HK_7         C    Lee      H3   \n",
       "40       I133 of HA2        M133 in HA2       HK_7        NA     NA      H3   \n",
       "44       K166 of HA1        V163 in HA1       HK_7         B    Lee      H3   \n",
       "45       E150 of HA2        E150 in HA2       HK_7        NA     NA      H3   \n",
       "48       S196 of HA1        S193 in HA1       HK_2         B  Wiley      H3   \n",
       "51       P130 of HA1        T128 in HA1       HK_8         B    Lee      H3   \n",
       "53       N266 of HA1        T262 in HA1       HK_8         E  Smith      H3   \n",
       "58       N154 of HA2        N154 in HA2       HK_2        NA     NA      H3   \n",
       "60       T276 of HA1        A272 in HA1       HK_1        NA     NA      H3   \n",
       "62       K311 of HA1        K307 in HA1       HK_1         C    Lee      H3   \n",
       "65       T251 of HA1        N248 in HA1   cali09_2        NA     NA      H1   \n",
       "68       P188 of HA1        P185 in HA1     cali09        NA     NA      H1   \n",
       "70       R211 of HA1        R208 in HA1     cali09      H1N1  Caton      H1   \n",
       "74       D145 of HA2        D145 in HA2   cali09_2        NA     NA      H1   \n",
       "79       T107 of HA2        T107 in HA2     cali09        NA     NA      H1   \n",
       "\n",
       "   H3_pos H3_seg  \n",
       "2      86    HA1  \n",
       "3      62    HA1  \n",
       "6       -      -  \n",
       "9     297    HA1  \n",
       "12    214    HA1  \n",
       "13      -      -  \n",
       "16    291    HA1  \n",
       "17    242    HA1  \n",
       "18    106    HA1  \n",
       "21     53    HA1  \n",
       "22    312    HA1  \n",
       "23    141    HA2  \n",
       "26    268    HA1  \n",
       "29    242    HA1  \n",
       "30    189    HA1  \n",
       "31    106    HA2  \n",
       "32      3    HA2  \n",
       "35    114    HA1  \n",
       "36    307    HA1  \n",
       "40    133    HA2  \n",
       "44    163    HA1  \n",
       "45    150    HA2  \n",
       "48    193    HA1  \n",
       "51    128    HA1  \n",
       "53    262    HA1  \n",
       "58    154    HA2  \n",
       "60    272    HA1  \n",
       "62    307    HA1  \n",
       "65    248    HA1  \n",
       "68    185    HA1  \n",
       "70    208    HA1  \n",
       "74    145    HA2  \n",
       "79    107    HA2  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HA_nonsyn[[\"SPECID\",\"onset\",\"mutation\",\"freq.var\",\"Ref_AA\",\"AA_pos\",\"Var_AA\",\"PDB_4JTV\",\"PDB_4HMG\",\"run\",\"Antigenic\",\"Source\",\"subtype\",\"H3_pos\",\"H3_seg\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Antigenic sites\n",
    "\n",
    "Here are the antigenic sites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "AA = HA_nonsyn.loc[HA_nonsyn[\"Antigenic\"]!=\"NA\"]\n",
    "len(AA)\n",
    "HA_nonsyn.to_csv(\"../data/processed/secondary/HA_nonsynom.csv\")\n",
    "AA[[\"HOUSE_ID\",\"ENROLLID\",\"SPECID\",\"onset\",\"mutation\",\"freq.var\",\"Ref_AA\",\"H3_pos\",\"Var_AA\",\"PDB_4HMG\",\"PDB_4JTV\",\"pcr_result\",\"Antigenic\",\"Source\",\"vaccination_status\",\"DPI\"]].to_csv(\"../data/processed/secondary/antigenic_isnv.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HOUSE_ID</th>\n",
       "      <th>ENROLLID</th>\n",
       "      <th>SPECID</th>\n",
       "      <th>onset</th>\n",
       "      <th>mutation</th>\n",
       "      <th>freq.var</th>\n",
       "      <th>Ref_AA</th>\n",
       "      <th>H3_pos</th>\n",
       "      <th>Var_AA</th>\n",
       "      <th>PDB_4HMG</th>\n",
       "      <th>PDB_4JTV</th>\n",
       "      <th>pcr_result</th>\n",
       "      <th>Antigenic</th>\n",
       "      <th>Source</th>\n",
       "      <th>vaccination_status</th>\n",
       "      <th>DPI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1302</td>\n",
       "      <td>301355</td>\n",
       "      <td>M54724</td>\n",
       "      <td>2011-03-20</td>\n",
       "      <td>HA_C329A</td>\n",
       "      <td>0.087608</td>\n",
       "      <td>['L']</td>\n",
       "      <td>86</td>\n",
       "      <td>['I']</td>\n",
       "      <td>L86 in HA1</td>\n",
       "      <td>Y84 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>E</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1111</td>\n",
       "      <td>300481</td>\n",
       "      <td>M54913</td>\n",
       "      <td>2011-03-30</td>\n",
       "      <td>HA_A258G</td>\n",
       "      <td>0.071215</td>\n",
       "      <td>['E']</td>\n",
       "      <td>62</td>\n",
       "      <td>['G']</td>\n",
       "      <td>I62 in HA1</td>\n",
       "      <td>G59 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>E</td>\n",
       "      <td>Smith</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2166</td>\n",
       "      <td>320661</td>\n",
       "      <td>MH0778</td>\n",
       "      <td>2012-02-13</td>\n",
       "      <td>HA_T963C</td>\n",
       "      <td>0.071359</td>\n",
       "      <td>['V']</td>\n",
       "      <td>297</td>\n",
       "      <td>['A']</td>\n",
       "      <td>V297 in HA1</td>\n",
       "      <td>I301 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3075</td>\n",
       "      <td>331045</td>\n",
       "      <td>MH2436</td>\n",
       "      <td>2012-12-10</td>\n",
       "      <td>HA_T718C</td>\n",
       "      <td>0.065651</td>\n",
       "      <td>['I']</td>\n",
       "      <td>214</td>\n",
       "      <td>['T']</td>\n",
       "      <td>I214 in HA1</td>\n",
       "      <td>K217 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>5147</td>\n",
       "      <td>50630</td>\n",
       "      <td>HS1294</td>\n",
       "      <td>2014-11-18</td>\n",
       "      <td>HA_A801T</td>\n",
       "      <td>0.163603</td>\n",
       "      <td>['I']</td>\n",
       "      <td>242</td>\n",
       "      <td>['L']</td>\n",
       "      <td>V242 in HA1</td>\n",
       "      <td>K245 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>5033</td>\n",
       "      <td>50141</td>\n",
       "      <td>HS1376</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_C236A</td>\n",
       "      <td>0.025191</td>\n",
       "      <td>['D']</td>\n",
       "      <td>53</td>\n",
       "      <td>['E']</td>\n",
       "      <td>N53 in HA1</td>\n",
       "      <td>K49 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>C</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>5033</td>\n",
       "      <td>50141</td>\n",
       "      <td>HS1376</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_A1011G</td>\n",
       "      <td>0.023458</td>\n",
       "      <td>['S']</td>\n",
       "      <td>312</td>\n",
       "      <td>['G']</td>\n",
       "      <td>N312 in HA1</td>\n",
       "      <td>T316 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>5269</td>\n",
       "      <td>51132</td>\n",
       "      <td>HS1425</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_T802C</td>\n",
       "      <td>0.028037</td>\n",
       "      <td>['I']</td>\n",
       "      <td>242</td>\n",
       "      <td>['T']</td>\n",
       "      <td>V242 in HA1</td>\n",
       "      <td>K245 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>D</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>5289</td>\n",
       "      <td>51220</td>\n",
       "      <td>HS1450</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>HA_G644T</td>\n",
       "      <td>0.037827</td>\n",
       "      <td>['K']</td>\n",
       "      <td>189</td>\n",
       "      <td>['N']</td>\n",
       "      <td>Q189 in HA1</td>\n",
       "      <td>A192 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>B</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>5034</td>\n",
       "      <td>50143</td>\n",
       "      <td>HS1604</td>\n",
       "      <td>2015-01-11</td>\n",
       "      <td>HA_G997G</td>\n",
       "      <td>0.118966</td>\n",
       "      <td>['I']</td>\n",
       "      <td>307</td>\n",
       "      <td>['R']</td>\n",
       "      <td>K307 in HA1</td>\n",
       "      <td>K311 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>5033</td>\n",
       "      <td>50141</td>\n",
       "      <td>MH7755</td>\n",
       "      <td>2014-12-03</td>\n",
       "      <td>HA_G564A</td>\n",
       "      <td>0.032187</td>\n",
       "      <td>['A']</td>\n",
       "      <td>163</td>\n",
       "      <td>['T']</td>\n",
       "      <td>V163 in HA1</td>\n",
       "      <td>K166 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>B</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>5219</td>\n",
       "      <td>50935</td>\n",
       "      <td>MH7851</td>\n",
       "      <td>2014-12-05</td>\n",
       "      <td>HA_T655C</td>\n",
       "      <td>0.174563</td>\n",
       "      <td>['F']</td>\n",
       "      <td>193</td>\n",
       "      <td>['S']</td>\n",
       "      <td>S193 in HA1</td>\n",
       "      <td>S196 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>B</td>\n",
       "      <td>Wiley</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>5263</td>\n",
       "      <td>51106</td>\n",
       "      <td>MH7860</td>\n",
       "      <td>2014-12-06</td>\n",
       "      <td>HA_A459G</td>\n",
       "      <td>0.110522</td>\n",
       "      <td>['T']</td>\n",
       "      <td>128</td>\n",
       "      <td>['A']</td>\n",
       "      <td>T128 in HA1</td>\n",
       "      <td>P130 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>B</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>5302</td>\n",
       "      <td>51273</td>\n",
       "      <td>MH8052</td>\n",
       "      <td>2014-12-13</td>\n",
       "      <td>HA_G862A</td>\n",
       "      <td>0.030123</td>\n",
       "      <td>['S']</td>\n",
       "      <td>262</td>\n",
       "      <td>['N']</td>\n",
       "      <td>T262 in HA1</td>\n",
       "      <td>N266 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>E</td>\n",
       "      <td>Smith</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>5034</td>\n",
       "      <td>50143</td>\n",
       "      <td>MH9045</td>\n",
       "      <td>2015-01-11</td>\n",
       "      <td>HA_G997G</td>\n",
       "      <td>0.160589</td>\n",
       "      <td>['I']</td>\n",
       "      <td>307</td>\n",
       "      <td>['R']</td>\n",
       "      <td>K307 in HA1</td>\n",
       "      <td>K311 of HA1</td>\n",
       "      <td>A/H3N2</td>\n",
       "      <td>C</td>\n",
       "      <td>Lee</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>4185</td>\n",
       "      <td>UM40738</td>\n",
       "      <td>MH5405</td>\n",
       "      <td>2013-12-14</td>\n",
       "      <td>HA_G685A</td>\n",
       "      <td>0.020723</td>\n",
       "      <td>['R']</td>\n",
       "      <td>208</td>\n",
       "      <td>['K']</td>\n",
       "      <td>R208 in HA1</td>\n",
       "      <td>R211 of HA1</td>\n",
       "      <td>A/H1N1</td>\n",
       "      <td>H1N1</td>\n",
       "      <td>Caton</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    HOUSE_ID ENROLLID  SPECID       onset   mutation  freq.var Ref_AA H3_pos  \\\n",
       "2       1302   301355  M54724  2011-03-20   HA_C329A  0.087608  ['L']     86   \n",
       "3       1111   300481  M54913  2011-03-30   HA_A258G  0.071215  ['E']     62   \n",
       "9       2166   320661  MH0778  2012-02-13   HA_T963C  0.071359  ['V']    297   \n",
       "12      3075   331045  MH2436  2012-12-10   HA_T718C  0.065651  ['I']    214   \n",
       "17      5147    50630  HS1294  2014-11-18   HA_A801T  0.163603  ['I']    242   \n",
       "21      5033    50141  HS1376  2014-12-03   HA_C236A  0.025191  ['D']     53   \n",
       "22      5033    50141  HS1376  2014-12-03  HA_A1011G  0.023458  ['S']    312   \n",
       "29      5269    51132  HS1425  2014-12-06   HA_T802C  0.028037  ['I']    242   \n",
       "30      5289    51220  HS1450  2014-12-13   HA_G644T  0.037827  ['K']    189   \n",
       "36      5034    50143  HS1604  2015-01-11   HA_G997G  0.118966  ['I']    307   \n",
       "44      5033    50141  MH7755  2014-12-03   HA_G564A  0.032187  ['A']    163   \n",
       "48      5219    50935  MH7851  2014-12-05   HA_T655C  0.174563  ['F']    193   \n",
       "51      5263    51106  MH7860  2014-12-06   HA_A459G  0.110522  ['T']    128   \n",
       "53      5302    51273  MH8052  2014-12-13   HA_G862A  0.030123  ['S']    262   \n",
       "62      5034    50143  MH9045  2015-01-11   HA_G997G  0.160589  ['I']    307   \n",
       "70      4185  UM40738  MH5405  2013-12-14   HA_G685A  0.020723  ['R']    208   \n",
       "\n",
       "   Var_AA      PDB_4HMG      PDB_4JTV pcr_result Antigenic Source  \\\n",
       "2   ['I']   L86 in HA1    Y84 of HA1      A/H3N2         E    Lee   \n",
       "3   ['G']   I62 in HA1    G59 of HA1      A/H3N2         E  Smith   \n",
       "9   ['A']  V297 in HA1   I301 of HA1      A/H3N2         C    Lee   \n",
       "12  ['T']  I214 in HA1   K217 of HA1      A/H3N2         D    Lee   \n",
       "17  ['L']  V242 in HA1   K245 of HA1      A/H3N2         D    Lee   \n",
       "21  ['E']   N53 in HA1    K49 of HA1      A/H3N2         C  Wiley   \n",
       "22  ['G']  N312 in HA1   T316 of HA1      A/H3N2         C    Lee   \n",
       "29  ['T']  V242 in HA1   K245 of HA1      A/H3N2         D    Lee   \n",
       "30  ['N']  Q189 in HA1   A192 of HA1      A/H3N2         B  Wiley   \n",
       "36  ['R']  K307 in HA1   K311 of HA1      A/H3N2         C    Lee   \n",
       "44  ['T']  V163 in HA1   K166 of HA1      A/H3N2         B    Lee   \n",
       "48  ['S']  S193 in HA1   S196 of HA1      A/H3N2         B  Wiley   \n",
       "51  ['A']  T128 in HA1   P130 of HA1      A/H3N2         B    Lee   \n",
       "53  ['N']  T262 in HA1   N266 of HA1      A/H3N2         E  Smith   \n",
       "62  ['R']  K307 in HA1   K311 of HA1      A/H3N2         C    Lee   \n",
       "70  ['K']  R208 in HA1   R211 of HA1      A/H1N1      H1N1  Caton   \n",
       "\n",
       "    vaccination_status  DPI  \n",
       "2                    1    1  \n",
       "3                    0    0  \n",
       "9                    1    1  \n",
       "12                   1    1  \n",
       "17                   1    1  \n",
       "21                   1    1  \n",
       "22                   1    1  \n",
       "29                   1    2  \n",
       "30                   1   -1  \n",
       "36                   1    1  \n",
       "44                   1    2  \n",
       "48                   0    3  \n",
       "51                   1    3  \n",
       "53                   1    0  \n",
       "62                   1    2  \n",
       "70                   0    2  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AA[[\"HOUSE_ID\",\"ENROLLID\",\"SPECID\",\"onset\",\"mutation\",\"freq.var\",\"Ref_AA\",\"H3_pos\",\"Var_AA\",\"PDB_4HMG\",\"PDB_4JTV\",\"pcr_result\",\"Antigenic\",\"Source\",\"vaccination_status\",\"DPI\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Here are the frequencies of the antigenic mutations against day post symptom onset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAuEAAAIACAYAAAAsWLK1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xl4FHWex/FPd9I5IJ2D0EBIBASSAAYREgUV5PB6xANQ\nVFAgAqIy7KrjqKPiOcios17MOj4rO45yeI7PKKIIzqLGUfEAdRRnwXBDOAyEnOTu2j+QXlsEG0l+\nVV28X8/D81DdVZ1P9Vf0Y/Hrao9lWZYAAAAAGOO1OwAAAABwrKGEAwAAAIZRwgEAAADDKOEAAACA\nYZRwAAAAwDBKOAAAAGBYrN0Bfqni4mItXbpUlmVpwIABGjx4cNjzX331lT788ENJUlxcnM4//3x1\n6tQpomMBAACA1hSVJTwYDGrJkiUqLCyU3+/X3LlzlZubq0AgENonLS1NkydPVkJCgoqLi7V48WJN\nmzYtomMBAACA1hSVy1FKSkqUnp6u1NRUxcTEKC8vT2vXrg3b57jjjlNCQoIkKSsrS1VVVREfCwAA\nALSmqCzhVVVVSk5ODm0nJyersrLykPt//vnn6tmz5y86FgAAAGhpUbkc5Uhs3LhRX375paZMmRLR\n/pWVlaqurg57LCkpKay4AwAAAEcjKku43+9XRUVFaLuysvInS/LOnTu1ePFiTZgwQYmJiREdu2rV\nKhUVFYW9ztChQzV8+PCWPg0AAAAco6KyhGdmZqqsrEzl5eVKSkrS6tWrNXbs2LB9ysvL9fLLL2vM\nmDFq165dxMfm5+crNzc37LWSkpK0d+9eNTU1te6JGRAfH6/6+nq7Yxy12NhYpaWluWYuErNxKubi\nTMzFmdwyF4nZONWBubhBVJZwr9erkSNHasGCBbIsS/3791cgENDKlSslSQUFBXr//fdVW1urN998\nM3TMNddcc8hjD0hOTv7Jq+qlpaVqbGw0c4KtKDY21hXncUBTU5NrzofZOBNzcSbm4kxum4vEbNB6\norKES1J2drays7PDHisoKAj9/qKLLtJFF10U8bEAAACAKVF5dxQAAAAgmlHCAQAAAMMo4QAAAIBh\nlHAAAADAMEo4AAAAYBglHAAAADCMEg4AAAAYRgkHAAAADKOEAwAAAIZRwgEAAADDKOEAAACAYZRw\nAAAAwDBKOAAAAGAYJRwAAAAwjBIOAAAAGEYJBwAAAAyjhAMAAACGUcIBAAAAwyjhAAAAgGGUcAAA\nAMAwSjgAAABgGCUcAAAAMIwSDgAAABhGCQcAAAAMo4QDAAAAhlHCAQAAAMMo4QAAAIBhlHAAAADA\nMEo4AAAAYBglHAAAADCMEg4AAAAYRgkHAAAADKOEAwAAAIZRwgEAAADDKOEAAACAYZRwAAAAwDBK\nOAAAAGAYJRwAAAAwjBIOAAAAGEYJBwAAAAyjhAMAAACGUcIBAAAAwyjhAAAAgGGUcAAAAMAwSjgA\nAABgGCUcAAAAMIwSDgAAABhGCQcAAAAMo4QDAAAAhlHCAQAAAMMo4QAAAD9QVVWlzz77TDt37rQ7\nClyMEg4AAPC9LVu2aPjw4Tr//PPVs2dP/eMf/7A7ElzKY1mWZXcIp6urq1NdXZ3c8FZ5vV4Fg0G7\nYxw1j8ejuLg4NTQ0uGIuErNxKubiTMzFmdwwl9tuu01z584NbQ8cOFBvvfWWjYlahhtmI+3/M5Oa\nmmp3jBYRa3eAaJCQkKCqqio1NjbaHeWoJSYmqra21u4YR83n8yk1NVU1NTWumIvEbJyKuTgTc3Em\nN8zlx0XV4/FE/TlJ7piNtP/PjFuwHAUAAOB71113nbp27SpJSklJ0cyZM21OBLfiSjgAAMD3MjIy\ntHz5cpWUlOjEE09UY2OjK/6WAs7DlXAAAIAfSExMVO/evV2z9hjORAkHAAAADKOEAwAAAIZRwgEA\nAADDKOEAAACAYZRwAAAAwDBKOAAAAGAYJRwAAAAwjBIOAAAAGEYJBwAAAAyjhAMAAACGUcIBAAAA\nwyjhAAAAgGGUcAAAAMAwSjgAAABgGCUcAAAAMIwSDgAAABhGCQcAAAAMo4QDAAAAhlHCAcDlVq1a\npcsuu0znnnuuVq5caXccAICkWLsDAABaz969ezVhwgRVVlZKkj755BN9+OGHSktLszkZABzbuBIO\nAC62devWUAGXpIqKCm3dutXGRAAAiRIOAK7WvXt3derUKbSdkZGh7t2725gIACCxHAUAXC0pKUmv\nvPKKnnrqKSUkJGjq1KlKSkqyOxYAHPMo4QDgcscff7weeeQRBQIBlZaWqrGx0e5IAHDMYzkKAAA2\nqqurszsCABtQwgEAsMEXX3yhvLw8JSYmatKkSWpoaLA7EgCDKOEAANjglltu0XfffSdJWrp0qZ5/\n/nmbEwEwiRIOAIANysvLD7sNwN0o4QAA2ODqq68O/T49PV0XX3yxjWkAmMbdUQAAsMF1112nk08+\nWRUVFerXr5/S09PtjgTAIEo4AAA2GTRoELeOBI5RLEcBAAAADKOEAwAAAIZRwgEAAADDKOEAAACA\nYZRwAAAAwDBKOAAAAGAYJRwAAAAwjBIOAAAAGEYJBwAAAAyjhAMAAACGUcIBAAAAwyjhAAAAgGGU\ncAAAAMAwSjgAAABgGCUcAAAAMCzW7gC/VHFxsZYuXSrLsjRgwAANHjw47Pndu3frtdde044dO3Tm\nmWfqtNNOCz332GOPKSEhQR6PR16vV9dcc43p+AAAADiGRWUJDwaDWrJkiQoLC+X3+zV37lzl5uYq\nEAiE9klMTNTIkSO1Zs2ag473eDy66qqrlJiYaDI2AAAAIClKl6OUlJQoPT1dqampiomJUV5entau\nXRu2T9u2bdW5c2d5vT99ipZlmYgKAAAAHCQqr4RXVVUpOTk5tJ2cnKySkpIjeo358+fL6/UqPz9f\n+fn5LR0RAAAAOKSoLOFHa+rUqfL7/aqpqdH8+fPVvn17de3aVZJUWVmp6urqsP2TkpIUG+uOtyom\nJkY+n8/uGEftwDzcMheJ2TgVc3Em5uJMbpmLxGycyi3zkKK0hPv9flVUVIS2Kysrw66MR3K8tH/J\nSu/evVVSUhIq4atWrVJRUVHY/pMmTVKnTp0UHx/fAunt55Z/gPft26f4+HjXzEViNk7FXJyJuTiT\nW+YiMRu0rqicRmZmpsrKylReXq6kpCStXr1aY8eOjejYhoYGWZal+Ph4NTQ0aP369Ro6dGjo+fz8\nfOXm5oYdk5SUpPr6ejU1NbXoedghPj5e9fX1dsc4arGxsUpLS9PevXtdMReJ2TgVc3Em5uJMbpmL\nxGycKjY2Vm3atLE7RouIyhLu9Xo1cuRILViwQJZlqX///goEAlq5cqUkqaCgQNXV1Zo7d67q6+vl\n8Xj08ccfa8aMGdq3b59efPFFeTweBYNB9e3bVz179gy9dnJy8k9eVS8tLVVjY6Oxc2wtsbGxrjiP\nA5qamlxzPszGmZiLMzEXZ3LbXCRmg9YTlSVckrKzs5WdnR32WEFBQej3SUlJuummmw46Lj4+XtOn\nT2/1fAAAAMChROUtCgEAAIBoRgkHAAAADKOEAwAAAIZRwgEAAADDKOEAAACAYZRwAAAAwDBKOAAA\nAGAYJRwAAAAwjBIOAAAAGEYJBwAAAAyjhAMAAACGUcIBAAAAwyjhAAAAgGGUcAAAAMAwSjgAAABg\nGCUcAAAAMIwSDgAAABhGCQcAAAAMo4QDAAAAhlHCAQAAAMMo4QAAAIBhlHAAAADAMEo4AAAAYBgl\nHAAAADCMEg4AAAAYRgkHAAAADKOEAwAAAIZRwgEAAADDKOEAAACAYZRwAAAAwDBKOAAAAGAYJRwA\nAAAwjBIOAAAAGEYJBwAAAAyjhAMAAACGUcIBAAAAwyjhAAAAgGGUcAAAAMAwSjgAAABgGCUcAAAA\nMIwSDgAAABhGCQcAAAAMo4QDAAAAhlHCAQAAAMMo4QAAAIBhlHAAAADAMEo4AAAAYBglHAAAADCM\nEg4AAAAYRgkHAAAADKOEAwAAAIZRwgEAAADDPJZlWXaHcLq6ujrV1dXJDW+V1+tVMBi0O8ZR83g8\niouLU0NDgyvmIjEbp2IuzsRcnMktc5GYjVN5PB6lpqbaHaNFxNodIBokJCSoqqpKjY2Ndkc5aomJ\niaqtrbU7xlHz+XxKTU1VTU2NK+YiMRunYi7OxFycyS1zkZiNU/l8PrsjtBiWowAAAACGUcIBAAAA\nwyjhAAAAgGGUcAAAAMAwSjgAAABgGCUcAAAAMIwSDgAAABhGCQcAAAAM48t6AOAYsGvXLjU1NSk2\nln/tA4ATcCUcAFzugQceUN++fdW5c2fNnj3b7jgAAFHCAcDVNmzYoCeeeCK0PWfOHG3YsMHGRAAA\niRIOAK7W0NAQ0WMAALMo4QDgYrm5ubroootC26NHj1avXr1sTAQAkPhgJgC4msfj0ZNPPqlp06Yp\nLS1N2dnZampqsjsWABzzKOEA4HIej0cDBw5UIBBQaWmp3XEAAGI5CgAAAGAcJRwAAAAwjBIOAAAA\nGEYJBwAAAAyjhAMAAACGUcIBAAAAwyjhAAAAgGGUcAAAAMAwSjgAAABgGCUcAAAAMIwSDgAAABhG\nCQcAAAAMo4QDAAD8wO7du7V8+XKtX7/e7ihwMUo4AADA99atW6fhw4dr/Pjx6tOnj5YtW2Z3JLgU\nJRwAAOB7zzzzjMrKyiRJDQ0Nevzxx21OBLeihAMAAHwvISEhbDsxMdGmJHA7SjgAAMD3fvWrX6l3\n796SpI4dO+q+++6zORHcKtbuAAAAAE6Rnp6uZcuWqaysTL169VJlZaUaGxvtjgUX4ko4AADAD8TE\nxKhz586Kj4+3OwpcjBIOAAAAGEYJBwAAAAyjhAMAAACGUcIBAAAAwyjhAAAAgGGUcAAAAMAwSjgA\nAABgGCUcAAAAMIwSDgAAABhGCQcAAAAMo4QDAAAAhlHCAQAAAMMo4QAAAIBhlHAAAADAsFi7A/xS\nxcXFWrp0qSzL0oABAzR48OCw53fv3q3XXntNO3bs0JlnnqnTTjst4mMBAACA1hSVJTwYDGrJkiUq\nLCyU3+/X3LlzlZubq0AgENonMTFRI0eO1Jo1a474WAAAAKA1ReVylJKSEqWnpys1NVUxMTHKy8vT\n2rVrw/Zp27atOnfuLK/Xe8THAgAAAK0pKkt4VVWVkpOTQ9vJycmqrKxs9WMBAACAlhCVy1FaU2Vl\npaqrq8MeS0pKUmysO96qmJgY+Xw+u2MctQPzcMtcJGbjVMzFmZiLM7llLhKzcSq3zEOK0hLu9/tV\nUVER2q6srAy7un00x65atUpFRUVhxwwdOlTDhw8/ytRoDWlpaXZHwCEwG2diLs7EXJyL2aC1RFzC\n58yZoyuvvFLt27dvzTwRyczMVFlZmcrLy5WUlKTVq1dr7NixLXJsfn6+cnNzw45JSkrS3r171dTU\n1KLnYYf4+HjV19fbHeOoxcbGKi0tzTVzkZiNUzEXZ2IuzuSWuUjMxqkOzMUNIi7h77zzjmbOnKlh\nw4Zp4sSJGj16tOLj41sz2yF5vV6NHDlSCxYskGVZ6t+/vwKBgFauXClJKigoUHV1tebOnav6+np5\nPB59/PHHmjFjhuLj43/y2AOSk5N/8qp6aWmpGhsbjZ1ja4mNjXXFeRzQ1NTkmvNhNs7EXJyJuTiT\n2+YiMRu0Ho9lWVakO+/Zs0cvvviiFi5cqDVr1uiSSy7RpEmTdMYZZ7RmRkdwSwlPTExUbW2t3TGO\nms/nUyAQcM1cJGbjVMzFmZiLM7llLhKzcaoDc3GDI7o7Snp6umbMmKEVK1aoqKhIn332mYYPH65u\n3bpp9uzZB32gEQAAAMDBjvgWhcuXL9fkyZM1bNgwdejQQfPmzdPChQv1xRdf6LzzzmuNjAAAAICr\nRLwm/Oabb9aLL76olJQUTZw4UV9//bUyMzNDzw8cONA1C+UBAACA1hRxCa+rq9Orr76qk08++Sef\n9/l8oQ9GAgAAADi0iEv47bffrjZt2oQ9tnfvXtXW1qpz586SpF69erVsOgAAAMCFIl4TPnr0aG3b\nti3ssW3btmnMmDEtHgoAAABws4hL+Nq1a9W3b9+wx/r27as1a9a0eCgAAADAzSIu4R06dNC6devC\nHlu3bp3S09NbPBQAAADgZhGX8ClTpuiSSy7RG2+8oX/9619avHixxo4dq6uvvro18wEAAACuE/EH\nM2+77Tb5fD7dfPPN2rp1q4477jhdffXVuummm1ozHwAAAOA6EZdwr9erW265Rbfccktr5gEAAABc\nL+ISLu3/cOY///nPg76efsqUKS0aCgAAAHCziEv473//e/3ud79Tv379wu4X7vF4KOEAAADAEYi4\nhD/++OP69NNPdeKJJ7ZmHgAAAMD1Ir47SmJiIt+ICQAAALSAiEv4rFmz9O///u/asWOHgsFg2C8A\nAAAAkYt4OcpVV10lSfrzn/8cesyyLHk8HjU3N7d4MAAAAMCtIi7hGzdubM0cAAAAwDEj4hLetWtX\nSVIwGNSuXbuUkZHRaqEAAAAAN4t4TXh5ebmuuOIKJSQkqGfPnpKk119/XXfeeWerhQMAAADcKOIS\nft111yklJUWbN29WXFycJOnUU0/VSy+91GrhAAAAADeKeDnK8uXLtX37dvl8Pnk8HklSIBDQd999\n12rhAAAtY+fOnWpsbJTP57M7CgBAR3AlPCUlRbt37w57bMuWLawNBwCHmz17tk488URlZmbqd7/7\nnd1xAAA6ghJ+9dVX65JLLtG7776rYDCoFStWqLCwUNddd11r5gMAHIX169frySefDG0/8cQTWr9+\nvY2JAADSESxH+e1vf6vExETNmDFDjY2NmjJliq699lrdcMMNrZkPAHAUGhsbI3oMAGBWxCXc4/Ho\nhhtuoHQDQBTp1auXRo8erddee02SdPHFF6tXr142pwIARFzC33nnnUM+N2LEiBYJAwBoeU888YSu\nueYapaWlqXv37mpqarI7EgAc8yIu4VOnTg3bLi0tVUNDg7KysrRhw4YWDwYAaBkej0cFBQUKBAIq\nLS21Ow4AQEfxtfXNzc26//775ff7WzwUAAAA4GYR3x3lx2JiYjRz5kz94Q9/aMk8AAAAgOv94hIu\nSX//+9/l9R7VSwAAAADHnIiXoxx33HGhb8qUpH379qmuri7s/rMAAAAAfl7EJXzhwoVh223btlVO\nTo6Sk5NbPBQAAADgZhGX8KFDh7ZmDgAAAOCYEXEJnzhxYthylEOZP3/+UQUCAAAA3C7iT1Wmpqbq\ntddeU3Nzs7KyshQMBrVo0SKlpqaqR48eoV8AAAAADi/iK+Hffvut3nzzTQ0ZMiT02AcffKBZs2Zp\n2bJlrRIOAAAAcKOIr4R//PHHGjRoUNhjAwcO1IoVK1o8FAAAAOBmEZfw/v3764477lBtba0kqba2\nVjNnztRJJ53UauEAAAAAN4q4hD/77LP68MMPlZKSoo4dOyolJUUffPCB5s2b15r5AAAAANeJeE14\nt27d9NFHH2nr1q3avn27MjIy1KVLl9bMBgAAALjSEX3n/J49e/Tee++pqKhIXbp00fbt27Vt27bW\nygYAAAC4UsQlvKioSLm5uXruuec0a9YsSVJxcbGmT5/eauEAAAAAN4q4hN9444166aWXtHTpUsXG\n7l/FMnDgQH366aetFg4AAABwo4hL+KZNm3TmmWdKUuibM+Pi4tTU1NQ6yQAAAACXiriE9+nT56Av\n5fmf//kf9e3bt8VDAQAAAG4W8d1RHnnkEV1wwQU6//zzVVtbq2uvvVaLFy/WokWLWjMfAAAA4DoR\nXwkfNGiQvvrqK51wwgmaMmWKjj/+eH366ac6+eSTWzMfAAAA4DoRXQlvbm7WmWeeqWXLlunWW29t\n7UwAAACAq0V0JTwmJkYbN25UMBhs7TwAAACA60W8HOWee+7R9OnTtXnzZjU3NysYDIZ+AQAAAIhc\nxB/MvPrqqyVJ8+fPD92i0LIseTweNTc3t046AAAAwIU8lmVZh9th586d6tSpkzZv3nzIfbp27dri\nwZykrq5OdXV1+pm3Kip4vV5X/O2Fx+NRXFycGhoaXDEXidk4FXNxJubiTG6Zi8RsnMrj8Sg1NdXu\nGC3iZ6+E5+TkqLKyMlS0L774Yv3tb39r9WBOkpCQoKqqKjU2Ntod5aglJiaqtrbW7hhHzefzKTU1\nVTU1Na6Yi8RsnIq5OBNzcSa3zEViNk7l8/nsjtBifnZN+I//7++9995rrSwAAADAMeFnS/iB9d8A\nAAAAWsbPLkdpamrSu+++G7oi/uNtSRoxYkTrJQQAAABc5mdLeIcOHTRlypTQdnp6eti2x+PRhg0b\nWicdAAAA4EI/W8I3bdpkIAYAN/jss8+0bNky9enTR2PHjrU7Dr739ttv64EHHlBMTIzuuOMO/vYS\nABwg4vuEA8DhfP7557r00ktDdxFYuXKlfv/739ucCt99952mT5+uuro6SdK0adP08ccfKxAI2JwM\nAI5tEX9jJgAcznvvvRd2G6+lS5famAYH7Nq1K1TApf3fe7Br1y4bEwEAJEo4gBZy/PHHh2337NnT\npiT4oZ49e4bNIjs7m9kAgAOwHAVAixgzZozWr1+vxYsXq0ePHnrwwQftjgTt/4KOV199VQsXLlTb\ntm116aWXKiEhwe5YAHDM+9mvrcd+paWlfGOWg/h8PgUCAdfMRWI2TsVcnIm5OJNb5iIxG6c6MBc3\nYDkKAAAAYBglHAAAADCMEg4AAAAYRgkHAAAADKOEAwAAAIZRwgEAAADDuE84ABwDdu3apaamJsXG\n8q99AHACroQDgMs98MAD6tu3rzp37qzZs2fbHQcAIEo4ALjahg0b9MQTT4S258yZow0bNtiYCAAg\nUcIBwNUaGhoiegwAYBYlHABcLDc3VxdddFFoe/To0erVq5eNiQAAEh/MBABX83g8evLJJzVt2jSl\npaUpOztbTU1NdscCgGMeJRwAXM7j8WjgwIEKBAIqLS21Ow4AQCxHAQAAAIyjhAMAAACGUcIBAAAA\nwyjhAAAAgGGUcAAAAMAwSjgAAABgGCUcAAAAMIwSDgAAABhGCQcAAAAMo4QDAAAAhlHCAQAAAMMo\n4QAAAIBhlHAAAADAMEo4AAAAYBglHAAAADCMEg4AAAAYRgkHAAAADKOEAwAAAIZRwgEAAADDKOEA\nAACAYZRwAAAAwDBKOAAAAGAYJRwAAAAwjBIOAAAAGEYJBwAAAAyjhAMAAACGxdod4JcqLi7W0qVL\nZVmWBgwYoMGDBx+0z5IlS7Ru3Tr5fD6NHj1aGRkZkqTHHntMCQkJ8ng88nq9uuaaa0zHBwAAwDEs\nKkt4MBjUkiVLVFhYKL/fr7lz5yo3N1eBQCC0T3Fxsfbu3avrr79e27Zt0xtvvKFp06ZJkjwej666\n6iolJibadQoAAAA4hkXlcpSSkhKlp6crNTVVMTExysvL09q1a8P2WbNmjfr16ydJysrKUn19vaqr\nq0PPW5ZlNDMAAABwQFReCa+qqlJycnJoOzk5WSUlJYfdx+/3q7KyUklJSZKk+fPny+v1Kj8/X/n5\n+WaCAwAAAIrSEn60pk6dKr/fr5qaGs2fP1/t27dX165dJUmVlZVhV8wlKSkpSbGx7nirYmJi5PP5\n7I5x1A7Mwy1zkZiNUzEXZ2IuzuSWuUjMxqncMg8pSku43+9XRUVFaLuysjLsqveBfSorK39yH7/f\nL0lq27atevfurZKSklAJX7VqlYqKisJea+jQoRo+fHirnAuOTlpamt0RcAjMxpmYizMxF+diNmgt\nUVnCMzMzVVZWpvLyciUlJWn16tUaO3Zs2D65ubn67LPPlJeXp61btyohIUFJSUlqaGiQZVmKj49X\nQ0OD1q9fr6FDh4aOy8/PV25ubthrJSUlae/evWpqajJyfq0pPj5e9fX1dsc4arGxsUpLS3PNXCRm\n41TMxZmYizO5ZS4Ss3GqA3Nxg6gs4V6vVyNHjtSCBQtkWZb69++vQCCglStXSpIKCgqUk5Oj4uJi\nzZkzR3FxcRo1apQkqaamRi+++KI8Ho+CwaD69u2rnj17hl47OTn5oKvqklRaWqrGxkYzJ9iKYmNj\nXXEeBzQ1NbnmfJiNMzEXZ2IuzuS2uUjMBq0nKku4JGVnZys7OzvssYKCgrDt888//6Dj0tLSNH36\n9FbNBgAAABxOVN6iEAAAAIhmlHAAAADAMEo4AAAAYBglHAAAADCMEg4AAAAYRgkHAAAADKOEAwAA\nAIZRwgEAAADDKOEAAACAYZRwAAAAwDBKOAAAAGAYJRwAAAAwjBIOAAAAGEYJBwAAAAyjhAMAAACG\nUcIBAAAAwyjhAAAAgGGUcAAAAMAwSjgAAABgGCUcAAAAMIwSjqhVV1dndwQAAIBfhBKOqLN+/Xqd\ncsopSkxM1KhRo1RVVWV3JAAAgCNCCUfUufvuu7Vp0yZJ0ooVK/Tkk0/aGwgAAOAIUcIRdSoqKsK2\ny8vLbUoCAADwy1DCEXUmT54sr3f/P7pt2rTR+PHjbU4EAABwZGLtDgAcqUsuuUS5ubnauXOnevXq\npaysLLsjAQAAHBFKOKJS//79FQgEVFpaqsbGRrvjAAAAHBGWowAAAACGUcIBAAAAwyjhAAAAgGGU\ncAAAAMAwSjgAAABgGCUcAAAAMIxbFAIAYJMlS5Zo7969GjRokLp06WJ3HAAGUcIBALDBH/7wB82Z\nM0eSlJSUpMWLFysnJ8fmVABMYTkKAAA2eOmll0K/r66u1ptvvmljGgCmUcIBALBBx44dD7sNwN0o\n4QAA2OCxxx5Tr1691LZtW40fP16XX3653ZEAGEQJB9BiHnvsMQ0ZMkTnn3++duzYYXccwNFyc3P1\n/vvvq7q6WnPmzFFMTIzdkQAYxAczAbSIRYsW6eGHH5YkrV27VpWVlWFrXgEAwP/jSjiAFrFu3bqw\n7eLiYpuSAADgfJRwAC1i2LBhio39/79cO+ecc2xMAwCAs7EcBUCLyM/P18svv6xly5apd+/euuyy\ny2RZlt2xAEfbvXu3du7cqXbt2snj8dgdB4BBlHAALWbgwIEaPHiwAoGASktL1djYaHckwLHefvtt\nTZ8+XXXtcnKgAAAUsElEQVR1dcrLy9Nf//pXJScn2x0LgCEsRwEAwAb33Xef6urqJEmrV6/Wc889\nZ3MiACZRwgEAsEEwGDzsNgB3o4QDAGCD22+/XT6fT5KUnZ2t8ePH25wIgEmsCQcAwAYXXXSRBg0a\npPr6enXq1ClUyAEcGyjhAADYJDMzkw8yA8colqMAgMu9//77Ovfcc3XKKaeoqKjI7jgAAHElHABc\nbc+ePZo6dar27dsnSSosLNTHH3+s9PR0m5MBwLGNK+EA4GIlJSWhAi5J+/bt0/bt221MBACQJI/F\nV9r9rLq6OtXV1bni2/+8Xq8rboPl8XgUFxenhoYGV8xFYjZOFe1z2bdvnwYPHqxNmzZJkrp166YP\nPvhAbdq0sTfYUYr2uRzAnxfnYjbO5PF4lJqaaneMFsFylAgkJCSoqqrKFR+aSUxMVG1trd0xjprP\n51NqaqpqampcMReJ2ThVtM/F4/Hoqquu0oMPPiiPx6OpU6fK4/FE9TlJ0T+XA/jz4lzMxpncdBch\nlqMAgItt375d999/v+rq6lRbW6v77ruP5SgA4ACUcABwsbKyMjU1NYW2m5qaVFZWZmMiAIBECQcA\nV8vJyVH//v1D2wMGDFBOTo6NiQAAEmvCAcDV4uLi9PLLL+u1116T3+/Xueeeq7i4OLtjAcAxjxIO\nAC7Xpk0bFRYW8s2MAOAgLEcBAMAGJSUluvTSS5WTk6MHH3zQ7jgADONKOAAANrjhhhu0YsUKSdKj\njz6q7t27a8yYMTanAmAKV8IBALDBhg0bwrY3btxoUxIAdqCEAwBgg3PPPTf0e5/PpxEjRtiYBoBp\nLEcBAMAG999/v/r06aM9e/Zo2LBhOumkk+yOBMAgSjgAADaIiYnRlClTuGsNcIxiOQoAAABgGCUc\nAAAAMIwSDgAAABhGCQcAAAAMo4QDAAAAhlHCAQAAAMMo4QAAAIBhlHAAAADAMEo4AAAAYBglHAAA\nADCMEg4AAAAYRgkHAAAADKOEAwAAAIZRwgEAAADDKOEAAACAYZRwAAAAwDBKOAAAAGAYJRwAAAAw\njBIOAAAAGEYJBwAAAAyjhAMAAACGUcIBAAAAwyjhAAAAgGGUcAAAAMAwSjgAAABgGCUcAAAAMIwS\nDgAAABhGCQcAAAAMo4QDAAAAhlHCAQAAAMMo4QAAAIBhlHAAAADAMEo4AAAAYBglHAAAADCMEg4A\nAAAYRgkHAAAADKOEAwAAAIZRwgEAAADDYu0O8EsVFxdr6dKlsixLAwYM0ODBgw/aZ8mSJVq3bp18\nPp9Gjx6tjIyMiI8FAAAAWktUXgkPBoNasmSJJk6cqBkzZujrr79WaWlp2D7FxcXau3evrr/+el14\n4YV64403Ij4WAAAAaE1RWcJLSkqUnp6u1NRUxcTEKC8vT2vXrg3bZ82aNerXr58kKSsrS/X19aqu\nro7oWAAAAKA1RWUJr6qqUnJycmg7OTlZlZWVEe0TybEAAABAa4raNeFHyrKsiParrKxUdXV12GNJ\nSUmKjXXHWxUTEyOfz2d3jKN2YB5umYvEbJyKuTgTc3Emt8xFYjZO5ZZ5SFFawv1+vyoqKkLblZWV\nYVe3D+zzwyvcB/Zpbm4+7LGrVq1SUVFR2Gt17dpVl1xyidLS0lr6VPALVVZW6t1331V+fj5zcRhm\n40zMxZmYi3MxG2f64Vx+3P2iTVQuR8nMzFRZWZnKy8vV1NSk1atXKzc3N2yf3Nxc/fOf/5Qkbd26\nVQkJCUpKSvrZY/Pz83XNNdeEfo0ZM0abN28+6Oo47FVdXa2ioiLm4kDMxpmYizMxF+diNs7kprlE\n5ZVwr9erkSNHasGCBbIsS/3791cgENDKlSslSQUFBcrJyVFxcbHmzJmjuLg4jRo16rDHHpCcnBz1\n/2cFAAAAZ4vKEi5J2dnZys7ODnusoKAgbPv888+P+FgAAADAlKhcjgIAAABEs5h77733XrtDOJll\nWYqLi1O3bt0UHx9vdxx8j7k4F7NxJubiTMzFuZiNM7lpLh4r0nv3HePefvttffvtt4qJiVG7du00\natQoJSQk2B3rmPfNN9/ovffe0+7duzVt2jR17tzZ7kjHtOLiYi1dulSWZWnAgAEaPHiw3ZEgadGi\nRfr222/Vtm1b/epXv7I7Dr5XUVGhV199VTU1NfJ4PBowYIAGDRpkd6xjXlNTk5555hk1NzerublZ\nubm5Ouuss+yOhe8Fg0HNnTtXycnJuuKKK+yOc1Sidk24aT169NBZZ50lr9erv//97/rggw/4Q+kA\nHTt21Lhx47R48WK7oxzzgsGglixZosLCQvn9fs2dO1e5ublhH3yGPU466SSdcsopevXVV+2Ogh/w\ner0699xzlZGRofr6es2dO1c9evTgz4zNYmNjVVhYqLi4OAWDQT399NPasmWLunTpYnc0SPrkk08U\nCARUX19vd5SjxprwCPXo0UNe7/63Kysri2/ZdIj27dsrPT3d7hiQVFJSovT0dKWmpiomJkZ5eXla\nu3at3bGg/d91kJiYaHcM/Ijf71dGRoYkKT4+Xu3bt1dVVZXNqSBJcXFxkvZfFbcsi7/5doiKigoV\nFxdrwIABdkdpEVwJ/wW++OIL5eXl2R0DcJSqqqqw23smJyerpKTExkRA9Ni7d6927typzMxMu6NA\n/7/koaysTAUFBerQoYPdkSBp2bJlOvvss11xFVyihIeZP3/+T978/cwzzwx9oc/777+vmJgYnXji\niabjHbMimQsARKv6+nq9/PLLOu+886L+g2Zu4fV6dd1116murk4LFizQpk2b1K1bN7tjHdMOfK4l\nIyNDGzdutDtOi6CE/8CkSZMO+/wXX3yh4uJiFRYWGkoE6efnAmfw+/2qqKgIbVdWVvLFV8DPaG5u\n1ssvv6x+/fqpV69edsfBjyQkJCgnJ0fbt2+nhNtsy5YtWrt2rYqLi9XU1KT6+nr97W9/08UXX2x3\ntF+MEh6h4uJiffTRR5o8ebJiY3nbgB/LzMxUWVmZysvLlZSUpNWrV2vs2LF2x8L3uBGWMy1atEiB\nQIC7ojhITU2NYmJilJCQoMbGRq1fv17Dhg2zO9Yx76yzzgrdEGPTpk366KOPorqAS9yiMGJ//OMf\n1dzcHPpwU1ZWli644AKbU+F///d/9dZbb2nfvn1KSEhQp06dNGHCBLtjHbN+eIvC/v37a8iQIXZH\ngqRXXnlFmzZtUm1trdq2bavhw4erf//+dsc65m3ZskXPPPOMOnToII/HI2n/Mju+0dleu3btCt1J\nyLIsnXjiiTr99NNtToUfOlDCo/0WhZRwAAAAwDBuUQgAAAAYRgkHAAAADKOEAwAAAIZRwgEAAADD\nKOEAAACAYZRwAAAAwDBKOAAAAGAYJRwAAAAwjBIOAAAAGEYJBwAAAAyjhAMAAACGUcIBAAAAwyjh\nAAAAgGGUcAAAAMAwSjgAAABgGCUcAAAAMIwSDgAAABhGCQcAAAAMo4QDAAAAhlHCAThSt27d1KZN\nG6WkpKhdu3YaPHiwnnrqKVmWZXe0iB1//PF655137I5xWPPmzdOQIUPsjtFqNm/eLK/Xq2AwaHcU\nAAhDCQfgSB6PR2+++aYqKiq0efNm3XbbbXrooYc0depUu6O5imVZ8ng8dsdoNQfOL5r+5w3AsYES\nDsCxDhQnv9+vCy64QC+99JLmzZunf/3rX5KkJUuWaMCAAUpJSVHXrl113333hY694IIL9Kc//Sns\n9fr166dFixYd9HMOXC397//+b2VmZiozM1OPPPJI6PmGhgbdeOONyszMVFZWln7961+rsbFRkrRn\nzx5deOGFSktLU3p6uoYOHSpJmjRpkrZs2aILL7xQycnJevjhhw/6uYc69uGHH9bYsWPD9r3++uv1\n61//WpI0fPhw3XXXXTr99NPl9/s1atQo7dmzRxMmTFBKSooGDhyoLVu2hI71er36z//8T/Xo0UMd\nOnTQrbfeKklas2aNpk+frhUrVsjv96tdu3aSpMrKSk2aNEkdOnTQ8ccfr9mzZ4dea968eRo8eLBu\nuukmpaWlKTs7Wx999JGeffZZdenSRZ06ddL8+fMPOdMdO3Zo1KhRSk9PV05Ojv785z+Hnrvvvvt0\n+eWXq7CwUMnJyerbt68+//zz0PMPPfSQsrKylJycrN69e+vdd9+VtP+fkwcffFA9e/ZUIBDQuHHj\nVF5eLkmh9zQ1NVXJycn65JNPDpkNAIyyAMCBunXrZi1fvvygx7t06WL913/9l2VZllVUVGStXr3a\nsizL+vrrr61OnTpZixYtsizLsl5++WVr4MCBoeO+/PJLq3379lZjY+NBr7lp0ybL4/FYV1xxhVVb\nW2t9/fXXViAQCP38u+66yzr11FOt3bt3W7t377ZOO+006+6777Ysy7Juv/12a/r06VZzc7PV1NRk\nffDBB2Hn8M477xzyHA917I4dO6ykpCSroqLCsizLampqsjp06GB98cUXlmVZ1rBhw6zs7Gxr48aN\nVmVlpdWnTx8rOzvbeuedd6zm5mZr0qRJ1pQpU0I/x+PxWCNGjLDKy8utrVu3Wjk5OdbTTz9tWZZl\nPfvss9aQIUPCck2cONEaPXq0VVNTY23atMnKycmx/vKXv4T29/l81rx586xgMGjdeeedVlZWlvVv\n//ZvVkNDg/X2229bfr/fqqmp+clzHjJkSGjfL7/80goEAta7775rWZZl3XvvvVZiYqK1dOlSKxgM\nWrfffrs1aNAgy7Isa+3atdZxxx1n7dy507Isy9q8ebO1YcMGy7Is6/HHH7dOPfVUa/v27VZDQ4N1\n3XXXWePHjw/N1uv1WsFg8JBzAAA7cCUcQFTp3LmzysrKJElnnHGGTjjhBElSXl6exo0bp6KiIknS\nRRddpOLiYq1fv16StHDhQl1++eWKjY095Gvfe++9SkhIUF5eniZPnqwXXnhBkvT888/rnnvuUXp6\nutLT03XPPfdowYIFkiSfz6cdO3Zo48aNiomJ0emnnx72mtZhlkEc6thOnTppyJAh+utf/ypJeuut\ntxQIBHTSSSeFjp08ebK6desmv9+v8847T9nZ2Ro+fLi8Xq8uvfRSffHFF2E/67bbblNKSoqysrJ0\n4403hs7tx4LBoF566SU9+OCDatOmjbp27arf/OY3ofOV9q91nzRpkjwejy6//HJt375d99xzj3w+\nn84++2zFxcVp3bp1B732tm3btGLFCj300EPy+Xzq16+frr766rAr54MHD9a5554rj8ejiRMn6quv\nvpIkxcTEqKGhQatXr1ZTU5O6dOmi448/XpL01FNPafbs2crIyJDP59Pdd9+tV155RcFgMPT+H24O\nAGAHSjiAqFJSUhJaNvHJJ59oxIgR6tChg1JTU/XUU09p9+7dkqT4+HhddtllWrhwoSzL0gsvvKCJ\nEyce8nU9Ho+ysrJC2127dtX27dslSdu3b1eXLl1+8rlbbrlFPXr00DnnnKOePXvqoYceivhcbr31\n1kMeO2nSJC1cuFCS9Nxzzx2UvWPHjqHfJyYmHrRdXV0dtv+hzu3Hdu/eHSq5P9y/pKTkkD9bktq3\nb3/Yny/tfx/btWunNm3aHPK1O3XqFPp9mzZtVFdXp2AwqB49eujxxx/Xvffeq44dO+qKK67Qzp07\nJe1fTjRmzBi1a9dO7dq1U58+feTz+bRr1y5Xr3cHEN0o4QCixmeffabt27eH7uZx5ZVXavTo0Sop\nKVF5ebmuvfbasCueB4rs8uXL1bZtWw0cOPCQr21ZlrZu3Rra3rJlizp37ixp/9X3zZs3h57bvHlz\n6LmkpCQ9/PDDWr9+vV5//XU9+uijobXKP1cA27Zte8hjR48era+++krffPON3njjDV155ZVH8lYd\n5FDn9uOM7du3l8/nO+h8MzMzj+rnS///txg1NTVhWSJ97XHjxukf//hHKNtvf/tbSVKXLl301ltv\nqaysTGVlZdq7d69qamqUkZFBCQfgWJRwAI5XVVWlN954Q+PHj9fEiRPVp08fSVJ1dbXS0tLk8/n0\n6aef6vnnnw877tRTT5XH49FvfvObw14FP2DWrFmqra3VN998o2eeeUbjxo2TJI0fP17333+/du/e\nrd27d2vWrFmh13vzzTdDS178fr9iY2MVExMjaf8V4w0bNhzy5/3UsV7v/n8tJyQk6OKLL9YVV1yh\ngQMHhl3J/iX+4z/+Q+Xl5dq6davmzJkTOreOHTtq27ZtoQ+aer1eXXbZZZo5c6aqq6u1efNmPfbY\nY4d9/yJd6pGVlaXTTjtNt99+u+rr6/XVV1/p6aefjui1v/32W7377rtqaGhQXFycEhMTQ+/Vtdde\nqzvuuCP0YdTS0lK9/vrrkqRAICCv1xt6nwHAKSjhABzrwgsvVEpKirp06aIHHnhAN998s/7yl7+E\nnn/yySd11113KSUlRffff78uv/zyg15j0qRJWr16tSZMmPCzP2/o0KHq2bOnzj77bN16660688wz\nJUl33nmnCgoKdOKJJ6pfv34qKCjQzJkzJUnFxcU666yz5Pf7dfrpp2vGjBk644wzJEm33367Zs2a\npXbt2unRRx896Of91LEH7uYhSYWFhfr66681adKksON+ydXdUaNGKT8/XwMGDNCFF16oKVOmSJJG\njBihE044QZ06dVKHDh0kSX/84x/Vpk0bde/eXWeccYYmTJigyZMnH/K1f5zncPleeOEFbdy4UZ07\nd9Yll1yiWbNmafjw4T/72vX19brtttsUCATUuXNnlZaW6oEHHpAk3XDDDRo1apTOOeccpaSk6LTT\nTtOnn34qaf/SmJkzZ+r0009Xu3btQo8DgN08Fp9WAeBiCxcu1Ny5c/X+++8fcp/Nmzere/fuamxs\nDF1ddYJt27apV69e2rlzp5KSkn7x63i9Xq1bt07du3dvwXQAgKPhnP/aAEAL27dvn/70pz/p2muv\n/dl9nXY9IhgM6uGHH9a4ceOOqoADAJyJEg7Ald5++2116NBBGRkZGj9+/M/u76QP8O3bt08pKSl6\n5513wr6A6Jdy0rkBAPZjOQoAAABgGFfCAQAAAMMo4QAAAIBhlHAAAADAMEo4AAAAYBglHAAAADCM\nEg4AAAAY9n/qEvieOP5whwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x10dcdd208>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "<ggplot: (282797965)>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ggplot(aes(x='DPI', y='freq.var'), data=AA) +\\\n",
    "    geom_point()  + ylab(\"Frequency\")+xlab(\"Day post symptom onset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16\n"
     ]
    }
   ],
   "source": [
    "AA_nonsynom_n=len(AA)\n",
    "print(AA_nonsynom_n)\n",
    "write_to_summary(\"Antigenic iSNV n:\",AA_nonsynom_n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "HA_nonsyn_n=len(HA_nonsyn)\n",
    "write_to_summary(\"NS HA n:\",HA_nonsyn_n)\n",
    "#HA_nonsyn_n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Proportion of snv in antigenic sites 0.027634 - 0.010363\n",
      "Propotion of infections with antigenic sites 0.070000 - 0.030000 \n"
     ]
    }
   ],
   "source": [
    "frac_isnv = \"Proportion of snv in antigenic sites %f - %f\" % (len(AA)/len(snv), (len(AA.loc[AA[\"Source\"]==\"Wiley\"])+len(AA.loc[AA[\"Source\"]==\"Smith\"])+len(AA.loc[AA[\"Source\"]==\"Caton\"]))/len(snv))\n",
    "\n",
    "print(frac_isnv)\n",
    "\n",
    "better = AA.loc[(AA[\"Source\"]==\"Wiley\")|(AA[\"Source\"]==\"Smith\")|(AA[\"Source\"]==\"Caton\")]\n",
    "frac_infection = \"Propotion of infections with antigenic sites %f - %f \" % (len(AA.ENROLLID.unique())/200,len(better.ENROLLID.unique())/200)\n",
    "\n",
    "\n",
    "print(frac_infection)\n",
    "\n",
    "write_to_summary(\"Antigenic iSNV %:\",frac_isnv)\n",
    "write_to_summary(\"Antigenic infection n:\",frac_infection)\n",
    "#len(AA)/len(snv.loc[snv[\"class_factor\"]==\"Nonsynonymous\"])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Amino acids for Nextflu\n",
    "\n",
    "\n",
    "This a little old code but it used regex to get the amino acid position and and the variant or ref amino acid.\n",
    "\n",
    "H3_name is the minor\n",
    "H3_name_major is the major amino acid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/ipykernel/__main__.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "H3=[]\n",
    "H1=[]\n",
    "H3_major= []\n",
    "H1_major = []\n",
    "pat = re.compile('\\w(\\d+) (in|of) (HA\\d)') \n",
    "for index, row in HA_nonsyn.iterrows():\n",
    "    H3_pos = pat.match(row.loc[\"PDB_4HMG\"])\n",
    "    H1_pos = pat.match(row.loc[\"PDB_4JTV\"])\n",
    "    var = row.loc[\"Var_AA\"].split(\"'\")[1]\n",
    "    ref = row.loc[\"Ref_AA\"].split(\"'\")[1]\n",
    "    if H3_pos:\n",
    "        H3.append(H3_pos.group(3)+\":\"+H3_pos.group(1)+var)\n",
    "        H3_major.append(H3_pos.group(3)+\":\"+H3_pos.group(1)+ref)\n",
    "    else:\n",
    "        H3.append(\"-\")\n",
    "        H3_major.append(\"-\")\n",
    "    if H1_pos:\n",
    "        H1.append(H1_pos.group(3)+\":\"+H1_pos.group(1)+var)\n",
    "        H1_major.append(H1_pos.group(3)+\":\"+H1_pos.group(1)+ref)\n",
    "    else:\n",
    "        H1.append(\"-\")\n",
    "        H1_major.append(\"-\")\n",
    "        \n",
    "        \n",
    "HA_nonsyn[\"H3_name\"] = H3\n",
    "HA_nonsyn[\"H3_name_major\"] = H3_major\n",
    "\n",
    "\n",
    "HA_nonsyn[\"H1_name\"] = H1\n",
    "HA_nonsyn[\"H1_name_major\"] = H1_major\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "HA_nonsyn.to_csv(\"../data/processed/secondary/minor_nonsynom.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Is  AA_pos  base 0 or base 1\n",
    "\n",
    "I expect a G at PB1 594 (base1) )(593-base 0)  in sample MH7800 HK_8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for seq_record in SeqIO.parse(\"../data/processed/HK_8/coding_fa/MH7800_A.removed.parsed.coding.fasta\", \"fasta\"):    \n",
    "        if seq_record.name == \"PB1\":\n",
    "            PB1=seq_record\n",
    "            PB1_prot = seq_record.seq.translate(to_stop=True)\n",
    "        if seq_record.name == \"NR\":\n",
    "            NR=seq_record\n",
    "            NR_prot = seq_record.seq.translate(to_stop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PB1_prot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PB1_prot[593]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NR_prot[160]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is is base 1."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sci-py3",
   "language": "python",
   "name": "sci-py3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
